---
title: "R Notebook"
output: html_notebook
---

## Load libraries
```{r}
library(tidyverse)
library(ggtree)
# library(velocyto.R)
library(Seurat)
library(ggplot2)
source("../utils/utilFxns.R")
```

## Prepare Velocyto Data
Read in velocyto matrices
```{r}
data_dir <- "/Users/joshhahn/Google Drive File Stream/My Drive/shekharlab_data/projects/Chicken/Processed/181027_HM2JGCCXY/"
Chicken1A <- read.loom.matrices(file = paste0(data_dir, "Chicken1A/outs/possorted_genome_bam_7ECLI.loom"), engine = "hdf5r")
Chicken1B <- read.loom.matrices(file = paste0(data_dir, "Chicken1B/outs/possorted_genome_bam_CH9WW.loom"), engine = "hdf5r")
Chicken1C <- read.loom.matrices(file = paste0(data_dir, "Chicken1C/outs/possorted_genome_bam_EAT51.loom"), engine = "hdf5r")
Chicken1D <- read.loom.matrices(file = paste0(data_dir, "Chicken1D/outs/possorted_genome_bam_38SN7.loom"), engine = "hdf5r")

data_dir <- "/Users/joshhahn/Google Drive File Stream/My Drive/shekharlab_data/projects/Chicken/Processed/181219_HW5W5CCXY/"
ChickendRGC1 <- read.loom.matrices(file = paste0(data_dir, "ChickendRGC1/outs/possorted_genome_bam_A7XOH.loom"), engine = "hdf5r")
ChickendRGC2 <- read.loom.matrices(file = paste0(data_dir, "ChickendRGC2/outs/possorted_genome_bam_D1Y9A.loom"), engine = "hdf5r")
ChickenvRGC1 <- read.loom.matrices(file = paste0(data_dir, "ChickenvRGC1/outs/possorted_genome_bam_9VZ3B.loom"), engine = "hdf5r")
ChickenvRGC2 <- read.loom.matrices(file = paste0(data_dir, "ChickenvRGC2/outs/possorted_genome_bam_PF5SZ.loom"), engine = "hdf5r")
```

Add introns and exons and combine matrices
```{r}
Chicken1A_tot <- Chicken1A$spliced + Chicken1A$unspliced
Chicken1B_tot <- Chicken1B$spliced + Chicken1B$unspliced
Chicken1C_tot <- Chicken1C$spliced + Chicken1C$unspliced
Chicken1D_tot <- Chicken1D$spliced + Chicken1D$unspliced

ChickendRGC1_tot <- ChickendRGC1$spliced + ChickendRGC1$unspliced
ChickendRGC2_tot <- ChickendRGC2$spliced + ChickendRGC2$unspliced
ChickenvRGC1_tot <- ChickenvRGC1$spliced + ChickenvRGC1$unspliced
ChickenvRGC2_tot <- ChickenvRGC2$spliced + ChickenvRGC2$unspliced

Chicken_mat <- cbind(Chicken1A_tot, Chicken1B_tot, Chicken1C_tot, Chicken1D_tot, ChickenvRGC1_tot, ChickenvRGC2_tot,ChickendRGC1_tot,ChickendRGC2_tot)
```

Create Seurat object
```{r}
Chicken <- CreateSeuratObject(Chicken_mat, names.delim = ":")

Chicken@meta.data[colnames(Chicken1A_tot), 'orig.file'] = "Chicken1A"
Chicken@meta.data[colnames(Chicken1B_tot), 'orig.file'] = "Chicken1B"
Chicken@meta.data[colnames(Chicken1C_tot), 'orig.file'] = "Chicken1C"
Chicken@meta.data[colnames(Chicken1D_tot), 'orig.file'] = "Chicken1D"
Chicken@meta.data[colnames(ChickenvRGC1_tot), 'orig.file'] = "ChickenvRGC1"
Chicken@meta.data[colnames(ChickenvRGC2_tot), 'orig.file'] = "ChickenvRGC2"
Chicken@meta.data[colnames(ChickendRGC1_tot), 'orig.file'] = "ChickendRGC1"
Chicken@meta.data[colnames(ChickendRGC2_tot), 'orig.file'] = "ChickendRGC2"
```

Initial quality control metrics
```{r}
VlnPlot(Chicken, features = "nCount_RNA", pt.size = 0)
VlnPlot(Chicken, features = "nFeature_RNA", pt.size = 0)
```


## Cell Class Annotation
Perform initial clustering
```{r}
Chicken <- ClusterSeurat(Chicken)
saveRDS(Chicken, file = "../Species_Objects/Chicken_initial.rds")
```

Visualize initial clusters
```{r}
Chicken <- readRDS("../Species_Objects/Chicken_initial.rds")
DimPlot(Chicken, label = TRUE)
DimPlot(Chicken, group.by = "orig.file", cells = sample(colnames(Chicken)))
VlnPlot(Chicken, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Chicken, "nFeature_RNA", pt.size = 0) + RotatedAxis()
```

Calculate DE genes for cluster annotation
```{r}
Chicken_genes <- FindAllMarkers(Chicken)
saveRDS(Chicken_genes, file = "../Marker_Genes/Chicken_genes_initial.rds")
```

Look for major retinal cell classes
```{r}
Idents(Chicken) <- "seurat_clusters"
Chicken <- DendroOrder(Chicken)
BC_markers=c("VSX1", "VSX2")
AC_markers=c("PAX6","SLC32A1","GAD1", "GAD2", "SLC6A9")
MG_markers=c("SLC1A3","RLBP1")
HC_markers=c("ONECUT1", "ONECUT2","ONECUT3")
Cone_markers=c("ARR3","RS1", "GNGT2")
Rod_markers=c("RHO", "PDE6B", "PDC")
RGC_markers= c("RBPMS", "POU4F1", "THY1")
OL_markers = c("OLIG2")

DotPlot(Chicken, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Cone_markers, Rod_markers, MG_markers, OL_markers)) + RotatedAxis()
```

Annotate clusters based on expression of cell class markers
```{r}
Chicken@meta.data$cell_class = "N/A"
Chicken@meta.data[WhichCells(Chicken, idents =c(0, 3,5,7,12,29,45)),]$cell_class = "RGC"
Chicken@meta.data[WhichCells(Chicken, idents = c(9,13,14,16,21:27,31,33,35,37,40,41)),]$cell_class = "BC"
Chicken@meta.data[WhichCells(Chicken, idents = c(6,8,18,20,28,30,32,34,38,39,43)),]$cell_class = "AC"
Chicken@meta.data[WhichCells(Chicken, idents = c(11,17)),]$cell_class = "HC"
Chicken@meta.data[WhichCells(Chicken, idents = c(4,10,19,36)),]$cell_class = "PR"
Chicken@meta.data[WhichCells(Chicken, idents = c(2,44)),]$cell_class = "MG"
Chicken@meta.data[WhichCells(Chicken, idents = c(42)),]$cell_class = "OL"


# Check cluster 1 and 15 for DE genes, otherwise remove them
Chicken_genes <- readRDS(file = "../Marker_Genes/Chicken_genes_initial.rds")
head(subset(Chicken_genes, cluster == 1))$gene
head(subset(Chicken_genes, cluster == 15))$gene

# Remove clusters 1 and 15 for low counts and lack of DE genes
idents.remove <- c(1, 15)
Chicken = subset(Chicken, idents = setdiff(levels(Idents(Chicken)), idents.remove))

DimPlot(Chicken, group.by = "cell_class")
DimPlot(Chicken, label = TRUE)

saveRDS(Chicken, "../Species_Objects/Chicken_initial.rds")
```

Plot cell type by file
```{r}
Chicken <- readRDS("../Species_Objects/Chicken_initial.rds")

pdf("../Figures/ChickenUMAP.pdf", w=5, h=4, useDingbats = FALSE)
DimPlot(Chicken, group.by = "cell_class", cols = c("red","orange","green","blue","black","purple", "plum")) + ggtitle("Chicken")
dev.off()

counts <- table(Chicken@meta.data$cell_class, Chicken@meta.data$orig.file)
counts
counts <- t(t(counts) / colSums(counts))
barplot(counts, legend = rownames(counts), col= rev(c("red","orange","yellow","white","green","blue", "black")))
```

Import labels from study
```{r}
Chicken <- readRDS("../Species_Objects/Chicken_initial.rds")
Chick_meta <- read.csv( "../Species_Reference/Chicken/Chick_retina_atlas_meta.csv")
meta <- data.frame(cluster = Chick_meta$Cluster, row.names = Chick_meta$NAME)

barcode <- function(string){
    code <- strsplit(string, c(":"))
    barcode <- strsplit(code[[1]][2], "x")
    return (paste0(barcode[[1]][1], "-1"))
}

new_names <- unlist(lapply(colnames(Chicken), FUN = barcode))
Chicken@meta.data$barcode <- paste0(Chicken@meta.data$orig.file, "_", new_names)

Chicken@meta.data$annotated <- meta[Chicken@meta.data$barcode,]

saveRDS(Chicken, "../Species_Objects/Chicken_initial.rds")
```

### Reformat cell class assignment
```{r}
Chicken <- readRDS("../Species_Objects/Chicken_initial.rds")
Chicken@meta.data$assigned <- Chicken@meta.data$cell_class
Idents(Chicken) <- "annotated"

Chicken@meta.data$cell_class <- "Other"
Chicken@meta.data[WhichCells(Chicken, idents = paste0("AC-", 1:59)), "cell_class"] <- "GabaAC"
Chicken@meta.data[WhichCells(Chicken, idents = paste0("AC-", c(48,20,31,58,37,39,2,15,1,29,22,47,10,21,34,42,6,13))), "cell_class"] <- "GlyAC"
Chicken@meta.data[WhichCells(Chicken, idents = paste0("BP-", 1:22)), "cell_class"] <- "BP"
Chicken@meta.data[WhichCells(Chicken, idents = paste0("HC-", 1:5)), "cell_class"] <- "HC"
Chicken@meta.data[WhichCells(Chicken, idents = paste0("MG-", 1:5)), "cell_class"] <- "MG"

# Leave oligodendrocytes annotated as other
# Chicken@meta.data[WhichCells(Chicken, idents = paste0("OG-", 1:5)), "cell_class"] <- "OG"
Chicken@meta.data[WhichCells(Chicken, idents = paste0("RGC-", 1:41)), "cell_class"] <- "RGC"

Chicken@meta.data[WhichCells(Chicken, idents = c("BlueCones", "DBCones1", "DBCones2", "DBCones3", "DevDB13Cones", "DevDB2Cones", "DevNonDBCones", "GreenCones", "RedCones", "VioletCones")), "cell_class"] <- "Cone"
Chicken@meta.data[WhichCells(Chicken, idents = c("DevRods", "Rods")), "cell_class"] <- "Rod"


saveRDS(Chicken, "../Species_Objects/Chicken_initial.rds")

```


## Chicken RGC
Segregate into RGCs
```{r}
Idents(Chicken) <- "annotated"
RGC_idents <- paste0(rep("RGC", 41), "-", 1:41)
Chicken_RGC <- subset(Chicken, cells = WhichCells(Chicken, idents = RGC_idents))
```

Run clustering pipeline to compare to annotated clusters
```{r}
Chicken_RGC <- ClusterSeurat(Chicken_RGC)
DimPlot(Chicken_RGC, label = TRUE)
DimPlot(Chicken_RGC, label = TRUE, group.by = "annotated")

saveRDS(Chicken_RGC, "../Species_Objects/ChickenRGC_v1.rds")
```


Calculate DE genes for each cluster
```{r}
Idents(Chicken_RGC) <- "annotated"
ChickenRGC_markers <- FindAllMarkers(Chicken_RGC, only.pos = TRUE, test.use = "MAST", assay = "RNA")
saveRDS(ChickenRGC_markers, "../Species_Markers/ChickenRGCmarkers.rds")
```

Convert UMAP and diagonal gene plots to files
```{r}
Chicken_RGC <- readRDS("../Species_Objects/ChickenRGC_v1.rds")
RGC_markers <- readRDS("../Species_Markers/ChickenRGCmarkers.rds")

Idents(Chicken_RGC) <- "annotated"
Chicken_RGC@meta.data$annotated_num <- substring(Chicken_RGC@meta.data$annotated, 5)


pdf("../Figures/Species_UMAPs/ChickenRGC.pdf", w=4, h=4, useDingbats = FALSE)
DimPlot(Chicken_RGC, label = TRUE, group.by = "annotated_num") + NoLegend()
dev.off()

markers <- PlotUniqueMarkers(Chicken_RGC, RGC_markers, edits = TRUE)
pdf("Figures/Species_MarkerPlots/ChickenRGC.pdf", , w=6, h=4, useDingbats = FALSE)

dev.off()

```


### Recluster and compare
```{r}
Chicken_RGC <- readRDS("../Species_Objects/ChickenRGC_v1.rds")
Chicken_RGC <- FindNeighbors(Chicken_RGC, dims = 1:30)
Chicken_RGC <- FindClusters(Chicken_RGC, resolution = 1.2)
DimPlot(Chicken_RGC, group.by = "RNA_snn_res.1.2", label = TRUE)
```

```{r}
Idents(Chicken_RGC) <- "RNA_snn_res.1.2"
markers <- FindAllMarkers(Chicken_RGC, assay="RNA", max.cells.per.ident = 200)
exp_markers <- subset(markers, pct.2 < .5)
e_top_markers <- TopMarkers(exp_markers, num_markers = 5)
top_markers <- TopMarkers(markers, num_markers = 5)
DotPlot(Chicken_RGC, features = top_markers, assay = "RNA") + RotatedAxis()
```

```{r}
Chicken_RGC@meta.data$reclustered <- Chicken_RGC@meta.data$RNA_snn_res.1.2
Chicken_RGC@meta.data$annotated <- droplevels(Chicken_RGC@meta.data$annotated)
re_table <- table(Chicken_RGC@meta.data$reclustered, Chicken_RGC@meta.data$annotated)
MakePrettyConfusionMatrix(re_table)
```

```{r}
Idents(Chicken_RGC) <- "RNA_snn_res.1.2"

clusters = c(15, 38, 44, 42)
features = c()
for (clus in clusters){
  markers <- FindMarkers(Chicken_RGC, ident.1 = clus)
  features = c(features, head(rownames(markers), 5))
}
DotPlot(Chicken_RGC, features = features, assay = "RNA", idents = clusters) + RotatedAxis()

clus1 = 34
clus2 = 43
markers <- FindMarkers(Chicken_RGC, ident.1 = clus1, ident.2 = clus2)
DotPlot(Chicken_RGC, features = head(rownames(markers), 20), assay = "RNA", idents = c(clus1, clus2)) + RotatedAxis()
DotPlot(Chicken_RGC, features = head(rownames(subset(markers, avg_log2FC > 0)), 10), assay = "RNA", idents = c(clus1, clus2)) + RotatedAxis()
DotPlot(Chicken_RGC, features = head(rownames(subset(markers, avg_log2FC < 0)), 10), assay = "RNA", idents = c(clus1, clus2)) + RotatedAxis()
```




## Chicken BCs
Segregate into BCs
```{r}
Idents(Chicken) <- "annotated"
BC_idents <- paste0(rep("BP", 22), "-", 1:22)
Chicken_BC <- subset(Chicken, cells = WhichCells(Chicken, idents = BC_idents))
```

Run clustering pipeline to compare to annotated clusters
```{r}
Chicken_BC <- ClusterSeurat(Chicken_BC)
DimPlot(Chicken_BC, label = TRUE)
DimPlot(Chicken_BC, label = TRUE, group.by = "annotated")

saveRDS(Chicken_BC, "../Species_Objects/ChickenBC_v1.rds")
```

Calculate DE genes for each cluster
```{r}
Idents(Chicken_BC) <- "annotated"
ChickenBC_markers <- FindAllMarkers(Chicken_BC, only.pos = TRUE, test.use = "MAST", assay = "RNA")
saveRDS(ChickenBC_markers, "../Species_Markers/ChickenBCmarkers.rds")
```

Convert UMAP and diagonal gene plots to files
```{r}
Chicken_BC <- readRDS("../Species_Objects/ChickenBC_v1.rds")
BC_markers <- readRDS("../Species_Markers/ChickenBCmarkers.rds")

Idents(Chicken_BC) <- "annotated"
Chicken_BC@meta.data$annotated_num <- substring(Chicken_BC@meta.data$annotated, 4)


pdf("../Figures/Species_UMAPs/ChickenBC.pdf", w=4, h=4, useDingbats = FALSE)
DimPlot(Chicken_BC, label = TRUE, group.by = "annotated_num") + NoLegend()
dev.off()

markers <- PlotUniqueMarkers(Chicken_BC, BC_markers, edits = TRUE)
pdf("Figures/Species_MarkerPlots/ChickenBC.pdf", , w=6, h=4, useDingbats = FALSE)
dev.off()
```

### Recluster and compare
```{r}
Chicken_BC <- readRDS("../Species_Objects/ChickenBC_v1.rds")
Chicken_BC <- FindNeighbors(Chicken_BC, dims = 1:30)
Chicken_BC <- FindClusters(Chicken_BC, resolution = 0.8)
DimPlot(Chicken_BC, group.by = "RNA_snn_res.1.2", label = TRUE)
```


```{r}
Chicken_BC@meta.data$reclustered <- Chicken_BC@meta.data$RNA_snn_res.1.2
Chicken_BC@meta.data$annotated <- droplevels(Chicken_BC@meta.data$annotated)
re_table <- table(Chicken_BC@meta.data$reclustered, Chicken_BC@meta.data$annotated)
MakePrettyConfusionMatrix(re_table)
```

```{r}
Idents(Chicken_BC) <- "RNA_snn_res.1.2"

clusters = c(15, 38, 44, 42)
features = c()
for (clus in clusters){
  markers <- FindMarkers(Chicken_BC, ident.1 = clus)
  features = c(features, head(rownames(markers), 5))
}
DotPlot(Chicken_BC, features = features, assay = "RNA", idents = clusters) + RotatedAxis()

clus1 = 16
clus2 = 23
markers <- FindMarkers(Chicken_BC, ident.1 = clus1, ident.2 = clus2)
DotPlot(Chicken_BC, features = head(rownames(markers), 20), assay = "RNA", idents = c(clus1, clus2)) + RotatedAxis()
DotPlot(Chicken_BC, features = head(rownames(subset(markers, avg_log2FC > 0)), 20), assay = "RNA", idents = c(clus1, clus2)) + RotatedAxis()
DotPlot(Chicken_BC, features = head(rownames(subset(markers, avg_log2FC < 0)), 20), assay = "RNA", idents = c(clus1, clus2)) + RotatedAxis()
```



## Chicken PRs
Segregate into rods
```{r}
Chicken <- readRDS("../Species_Objects/Chicken_initial.rds")
Idents(Chicken) <- "annotated"
Chicken_Rods <- subset(Chicken, cells = WhichCells(Chicken, idents = c("Rods", "DevRods")))

Chicken_Rods <- ClusterSeurat(Chicken_Rods)

saveRDS(Chicken_Rods, "../Species_Objects/ChickenRods_v1.rds")
```

## Horizontal Cells
```{r}
Chicken <- readRDS("~Downloads/Species_Initial/Tree_shrew_initial.rds")

# Only keep HCs
Idents(Chicken) <- "cell_class"
Chicken_HC <- subset(Chicken, cells = WhichCells(Chicken, idents ="HC"))

# Remove cells with low and high counts or features
VlnPlot(Chicken_HC, "nCount_RNA", group.by = "orig.file", pt.size = .1)
VlnPlot(Chicken_HC, "nFeature_RNA", group.by = "orig.file", pt.size = .1)
Chicken_HC <- subset(Chicken_HC, subset = nCount_RNA < 6500)
Chicken_HC <- ClusterSeurat(Chicken_HC,  numPCs = 15, integrate.by = "orig.file")

DimPlot(Chicken_HC, label = TRUE)
DimPlot(Chicken_HC, group.by = "orig.file", cells = sample(colnames(Chicken_HC)))
VlnPlot(Chicken_HC, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Chicken_HC, "nFeature_RNA", pt.size = 0) + RotatedAxis()

DotPlot(Chicken_HC, features = HC_markers, assay = "RNA") + RotatedAxis()
DotPlot(Chicken_HC, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Cone_markers, Rod_markers, MG_markers, Other_markers), assay = "RNA") + RotatedAxis()
#1, 2, 5 are HC for sure
```


```{r}
Chicken <- readRDS("../Species_Objects/Chicken_initial.rds")
DE_genes <- FindMarkers(Chicken, ident.1 = c("Rods", "DevRods"))

BC_idents <- paste0(rep("BP", 22), "-", 1:22)
BC_genes <- FindMarkers(Chicken, ident.1 = BC_idents)
```

