---
title: "R Notebook"
output: html_notebook
---

# Preprocessing 

Load libraries
```{r}
library(reshape2)
library(Seurat)
library(SeuratDisk)
library(ggplot2)
source("../utils/utilFxns.R")
source("../utils/plottingFxns.R")
```

Set data directory and load loom files
```{r}
data_dir <- "/Users/joshhahn/Google Drive/My Drive/shekharlab_data/projects/Lamprey/"

Convert(paste0(data_dir, "Lamprey9/velo_outs/full.h5ad"), dest = "h5seurat")
Lamp9 <- LoadH5Seurat(paste0(data_dir, "Lamprey9/velo_outs/full.h5seurat"))

Convert(paste0(data_dir, "Lamprey10/velo_outs/full.h5ad"), dest = "h5seurat")
Lamp10 <- LoadH5Seurat(paste0(data_dir, "Lamprey10/velo_outs/full.h5seurat"))

```

Combine matrices and create new Seurat object
```{r}
Lamprey_mat <- cbind(Lamp9@assays$matrix@counts, Lamp10@assays$matrix@counts)
Lamprey <- CreateSeuratObject(Lamprey_mat, names.delim = ":")

Lamprey@meta.data[colnames(Lamp9), 'orig.file'] = "Lamprey9"
Lamprey@meta.data[colnames(Lamp10), 'orig.file'] = "Lamprey10"

```

```{r}
VlnPlot(Lamprey, features = "nCount_RNA", pt.size = 0)
VlnPlot(Lamprey, features = "nFeature_RNA", pt.size = 0)
```

```{r}
saveRDS(Lamprey, file = "../Species_Objects/Lamprey_initial.rds")
```

# Cell class annotation

```{r}
Lamprey <- readRDS("../Species_Objects/Lamprey_initial.rds")
Lamprey <- ClusterSeurat(Lamprey, numPCs = 20)
saveRDS(Lamprey, file = "../Species_Objects/Lamprey_initial.rds")
```


Visualize initial clusters
```{r}
DimPlot(Lamprey, label = TRUE)
DimPlot(Lamprey, group.by = "orig.file")
VlnPlot(Lamprey, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Lamprey, "nFeature_RNA", pt.size = 0) + RotatedAxis()
```


Look for major retinal cell classes
```{r}
# Lamprey <- DendroOrder(Lamprey)

RGC_markers= c("RBPMS","RBPMS2", "SLC17A6", "POU6F2", "THY1", "NEFL", "NEFM", "SNCG")
BC_markers=c("VSX2", "OTX2", "GRM6", "TRPM1", "VSX1", "CABP5", "GRIK1", "PRKCA")
AC_markers=c("TFAP2A", "TFAP2B", "TFAP2C", "GAD1", "GAD2", "SLC6A9", "C1QL1", "C1QL2")
HC_markers=c("ONECUT1", "LHX1", "CALB1", "TPM3")
Cone_markers=c("PDE6H", "CRX", "ARR3")
Rod_markers=c("SAG", "PDC", "RHO")
MG_markers=c("SLC1A3","RLBP1", "APOE")
Other_markers=c("S100B", "GFAP", "C1QA", "C1QB", "CLDN5", "IGFBP7")

# LOC116941015 is POU6F2
RGC_markers= c("LOC116941015", "RBPMS", "SLC17A6" )
# LOC116939504 is VSX2
# MSTRG.6810 is OTX2
# MSTRG.11002 is PRKCB, analogous to prkca?
BC_markers=c("LOC116939504", "MSTRG.6810", "MSTRG.11002")
# MSTRG.3969  LOC116946616 are variants of TFAP2B
AC_markers=c("TFAP2D", "MSTRG.3969", "LOC116946616","LOC103091742",  "GAD1", "SLC6A9")
HC_markers=c("ONECUT1", "LHX1")
# MSTRG.8796 is PDE6H, 
Cone_markers=c("MSTRG.8796","MSTRG.9291")
# MSTRG.9291 is a red opsin
# MSTRG.9293 is RHO
Rod_markers=c("MSTRG.9293")

MG_markers=c("APOB")
Other_markers=c("LOC116952989", "IGFBP7")



DotPlot(Lamprey, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Cone_markers, Rod_markers, MG_markers, Other_markers), group.by = "dendro_order") + RotatedAxis()


```

```{r}
plot(Lamprey@tools$BuildClusterTree)
```

Annotate clusters using DE genes and Dot plot of marker genes

```{r}
# Drop cluster 0 due to low counts
Lamprey <- DropClusters(Lamprey, idents = 0, refactor = FALSE)

Lamprey@meta.data$cell_class = "Not annotated"
Lamprey@meta.data[WhichCells(Lamprey, idents =c(30,8,13,5,28,16)),]$cell_class = "RGC"
Lamprey@meta.data[WhichCells(Lamprey, idents = c(9)),]$cell_class = "BP"
Lamprey@meta.data[WhichCells(Lamprey, idents = c(18,21,23,4, 17,1,10)),]$cell_class = "GlyAC"
Lamprey@meta.data[WhichCells(Lamprey, idents = c(24, 2,25, 14, 7, 3, 11, 19,26 )),]$cell_class = "GabaAC"
Lamprey@meta.data[WhichCells(Lamprey, idents = c(15,22)),]$cell_class = "HC"
Lamprey@meta.data[WhichCells(Lamprey, idents = c(34)),]$cell_class = "MG"
Lamprey@meta.data[WhichCells(Lamprey, idents = c(6, 33, 12, 20,31,29,27)),]$cell_class = "Other"

# Photoreceptors do separate into rods and cones, do a round of clustering to separate rods and cones
PR <- subset(Lamprey, cells = WhichCells(Lamprey, idents =  c(32)))
PR <- ClusterSeurat(PR)
Cone_markers=c("MSTRG.8796","MSTRG.9291")
Rod_markers=c("MSTRG.9293")
DotPlot(PR, features = c(Rod_markers, Cone_markers)) + RotatedAxis()
rod_cells <- WhichCells(PR, idents = 0)
cone_cells <- WhichCells(PR, idents = 1)

Lamprey@meta.data[rod_cells,]$cell_class = "Rod"
Lamprey@meta.data[cone_cells,]$cell_class = "Cone"

```

```{r}
DimPlot(Lamprey, group.by = "cell_class")
counts <- table(Lamprey@meta.data$cell_class, Lamprey@meta.data$orig.file)
counts
saveRDS(Lamprey, file = "../Species_Objects/Lamprey_initial.rds")
saveRDS(Lamprey@meta.data, file = "../Metadata//Lamprey_metadata.rds")

```


Investigate unmarked clusters to see if they have any unique markers
```{r}
Lamprey_genes <- readRDS("../Species_Markers/Lamprey_genes_initial.rds")

# 
markers <- FindMarkers(Lamprey, ident.1 = "")
Lamprey@meta.data[WhichCells(Lamprey, idents = c()),]$cell_class = "Cone"

```


```{r}
pdf("../Figures/Class Markers/Lamprey_all_classes.pdf", w=10, h=5, useDingbats = FALSE)
DotPlot(Lamprey, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Cone_markers, Rod_markers, MG_markers, Other_markers), group.by = "cell_class") + RotatedAxis()
dev.off()

```


# RGC clustering analysis
Create separate RGC object
```{r}
Lamprey <- readRDS("../Species_Objects/Lamprey_initial.rds")

# Only keep RGCs
Idents(Lamprey) <- "cell_class"
Lamprey_RGC <- subset(Lamprey, cells = WhichCells(Lamprey, idents ="RGC"))

# Remove cells with low and high counts or features
VlnPlot(Lamprey_RGC, "nCount_RNA", group.by = "orig.file", pt.size = .1)
VlnPlot(Lamprey_RGC, "nFeature_RNA", group.by = "orig.file", pt.size = .1)
```


```{r}
Lamprey_RGC <- subset(Lamprey_RGC, subset = nFeature_RNA > 500 & nCount_RNA < 40000)
```

## Run clustering pipeline

```{r}
Lamprey_RGC <- ClusterSeurat(Lamprey_RGC, integrate.by = "orig.file", numPCs = 20)
saveRDS(Lamprey_RGC, file = "../Species_Objects/LampreyRGC_v1.rds")
```

Visualize initial clusters
```{r}
# Lamprey_RGC <- readRDS("~/Species_Objects/LampreyRGC_integrated_v1.rds")
DimPlot(Lamprey_RGC, label = TRUE)
DimPlot(Lamprey_RGC, group.by = "orig.file", cells = sample(colnames(Lamprey_RGC)))
VlnPlot(Lamprey_RGC, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Lamprey_RGC, "nFeature_RNA", pt.size = 0) + RotatedAxis()
```

## Refine clusters
Test different resolution parameters
```{r}
DefaultAssay(Lamprey_RGC) <- "integrated"
Lamprey_RGC <- FindClusters(Lamprey_RGC, resolution = 0.8)
Lamprey_RGC <- FindClusters(Lamprey_RGC, resolution = 1.1)

DimPlot(Lamprey_RGC, label = TRUE, group.by = "integrated_snn_res.0.5")
DimPlot(Lamprey_RGC, label = TRUE, group.by = "integrated_snn_res.0.8")
DimPlot(Lamprey_RGC, label = TRUE, group.by = "integrated_snn_res.1.1")
```

Remove low quality clusters
```{r}
Idents(Lamprey_RGC) <- "integrated_snn_res.1.1"
VlnPlot(Lamprey_RGC, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Lamprey_RGC, "nFeature_RNA", pt.size = 0) + RotatedAxis()

# Remove clusters 0 and 19 do to low counts and features
Lamprey_RGC@meta.data$seurat_clusters <- Lamprey_RGC@meta.data$integrated_snn_res.1.1
Lamprey_RGC <- DropClusters(Lamprey_RGC, idents = c(0,19), refactor = FALSE)
```

Merge clusters
```{r}
DefaultAssay(Lamprey_RGC) <- "RNA"

Lamprey_markers <- FindAllMarkers(Lamprey_RGC, assay = "RNA", only.pos = TRUE)

DotPlot(Lamprey_RGC, features = TopMarkers(Lamprey_markers, num_markers = 3)) + RotatedAxis()

# 3,2 - primarily difference in expression levels
Lamprey_RGC <- MergeClusters(Lamprey_RGC, idents = c(2,3), refactor = FALSE)

# 11, 12 - some differentiating markers, keep separate

# 1,7 - share most markers
Lamprey_RGC <- MergeClusters(Lamprey_RGC, idents = c(1,7), refactor = TRUE)

# 9, 10 - separate


ident.1 = 9
ident.2 = 10
mark_dif <- FindMarkers(Lamprey_RGC, ident.1 = ident.1, ident.2 = ident.2)
mark_dif$pct.df <- mark_dif$pct.1 - mark_dif$pct.2
mark_dif <- mark_dif[order(mark_dif$pct.df),]

DotPlot(Lamprey_RGC, idents = c(ident.1, ident.2), features = c(head(rownames(mark_dif)), tail(rownames(mark_dif)))) + RotatedAxis()

mark.1 <- FindMarkers(Lamprey_RGC, ident.1 = ident.1, max.cells.per.ident = 1000, only.pos = TRUE)
mark.2 <- FindMarkers(Lamprey_RGC, ident.1 = ident.2, max.cells.per.ident = 1000, only.pos = TRUE)
DotPlot(Lamprey_RGC, features = unique(c(head(rownames(mark.1)), head(rownames(mark.2))))) + RotatedAxis()
```

## Summary Plots
```{r}
DefaultAssay(Lamprey_RGC) <- "RNA"
Lamprey_RGC@meta.data$type <- Lamprey_RGC@meta.data$seurat_clusters
DimPlot(Lamprey_RGC, label = TRUE, group.by = "type")
saveRDS(Lamprey_RGC, "../Species_Objects/LampreyRGC_int_v1.rds")
saveRDS(Lamprey_RGC@meta.data, "../Metadata/RGC/LampreyRGC_int_v1_metadata.rds")
```

```{r}
Idents(Lamprey_RGC) <- "type"
Lamprey_RGC <- DendroOrder(Lamprey_RGC)

Idents(Lamprey_RGC) <- "dendro_order"
Lamprey_markers <- FindAllMarkers(Lamprey_RGC, only.pos = TRUE, max.cells.per.ident = 1000)

PlotUniqueMarkers(Lamprey_RGC, Lamprey_markers)

DotPlot(Lamprey_RGC, features = TopMarkers(Lamprey_markers, num_markers = 2), group.by = "dendro_order") + RotatedAxis()
```


# BC clustering analysis
Create separate BC object
```{r}
Lamprey <- readRDS("../Species_Objects/Lamprey_initial.rds")

# Only keep BCs
Idents(Lamprey) <- "cell_class"
Lamprey_BC <- subset(Lamprey, cells = WhichCells(Lamprey, idents ="BP"))

# Remove cells with low and high counts or features
VlnPlot(Lamprey_BC, "nCount_RNA", group.by = "orig.file", pt.size = .1)
VlnPlot(Lamprey_BC, "nFeature_RNA", group.by = "orig.file", pt.size = .1)
```


```{r}
Lamprey_BC <- subset(Lamprey_BC, subset = nFeature_RNA > 500 & nCount_RNA < 40000)
```

## Run clustering pipeline

```{r}
Lamprey_BC <- ClusterSeurat(Lamprey_BC, integrate.by = "orig.file", numPCs = 20)
saveRDS(Lamprey_BC, file = "../Species_Objects/LampreyBC_v1.rds")
```

Visualize initial clusters
```{r}
# Lamprey_BC <- readRDS("~/Species_Objects/LampreyBC_integrated_v1.rds")
DimPlot(Lamprey_BC, label = TRUE)
DimPlot(Lamprey_BC, group.by = "orig.file", cells = sample(colnames(Lamprey_BC)))
VlnPlot(Lamprey_BC, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Lamprey_BC, "nFeature_RNA", pt.size = 0) + RotatedAxis()
```

## Refine clusters
Test different resolution parameters
```{r}
DefaultAssay(Lamprey_BC) <- "integrated"
Lamprey_BC <- FindClusters(Lamprey_BC, resolution = 0.8)
Lamprey_BC <- FindClusters(Lamprey_BC, resolution = 1.1)

DimPlot(Lamprey_BC, label = TRUE, group.by = "integrated_snn_res.0.5")
DimPlot(Lamprey_BC, label = TRUE, group.by = "integrated_snn_res.0.8")
DimPlot(Lamprey_BC, label = TRUE, group.by = "integrated_snn_res.1.1")
```

Remove cluster 1, 8 since it doesn't express BC markers and has low counts
```{r}
Idents(Lamprey_BC) <- "integrated_snn_res.0.8"
Lamprey_BC@meta.data$seurat_clusters <- Lamprey_BC@meta.data$integrated_snn_res.0.8
DefaultAssay(Lamprey_BC) <- "RNA"

VlnPlot(Lamprey_BC, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Lamprey_BC, "nFeature_RNA", pt.size = 0) + RotatedAxis()

BC_markers=c("LOC116939504", "MSTRG.6810", "MSTRG.11002")
DotPlot(Lamprey_BC, features = BC_markers, assay = "RNA") + RotatedAxis()

```

```{r}
Lamprey_BC <- DropClusters(Lamprey_BC, idents = c(1,8), refactor = TRUE)
```


```{r}
DefaultAssay(Lamprey_BC) <- "RNA"
Lamprey_BC <- DendroOrder(Lamprey_BC)

Rod <- c("MSTRG.11002")

Off <- c("GRIK2")
On <- c("ISL1")

DotPlot(Lamprey_BC, features = c(Rod, Off, On), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

Lamprey_BC@meta.data$on_off <- factor(Lamprey_BC@meta.data$seurat_clusters, levels = rev(c(4,8,3,6,7,10,11,13,15, 0, 1,2,5,9,12,14,16)))

DotPlot(Lamprey_BC, features = c(pan_BC, Rod, Cone, Off, On), group.by = "on_off", assay = "RNA") + RotatedAxis()
```

## Summary Plots
```{r}
DefaultAssay(Lamprey_BC) <- "RNA"
Lamprey_BC@meta.data$type <- Lamprey_BC@meta.data$seurat_clusters
DimPlot(Lamprey_BC, label = TRUE, group.by = "type")
saveRDS(Lamprey_BC, "../Species_Objects/LampreyBC_int_v1.rds")
saveRDS(Lamprey_BC@meta.data, "../Metadata/BC/LampreyBC_int_v1_metadata.rds")

```

# Import names from annotation
```{r}
Lamprey <- readRDS("../Species_Initial/Lamprey_initial.rds")

genes <- read.csv("../Orthology/GeneAnotation_Lamprey.csv")
colnames(genes) <- c("ID", "AltID", "Gene")

# Remove white space
genes$ID <- gsub(" ", "", genes$ID)
genes$AltID <- gsub(" ", "", genes$AltID)
genes$Gene <- gsub(" ", "", genes$Gene)

# If AltID and ID are both present, then they are the same
common_ID <- (genes$ID %in% rownames(Lamprey)) & (genes$AltID %in% rownames(Lamprey))
genes[common_ID, "ID"] != genes[common_ID, "AltID"]

# Replace ID with AltID if AltID is present, but not ID
genes[(genes$AltID %in% rownames(Lamprey)) & !(genes$ID %in% rownames(Lamprey)), "ID"] <- genes[(genes$AltID %in% rownames(Lamprey)) & !(genes$ID %in% rownames(Lamprey)), "AltID"]
genes$AltID <- NULL

# Remove genes not present and duplicate rows


genes <- genes[genes$ID %in% rownames(Lamprey) , ]

genes <- genes[!duplicated(genes),]

```

## Initial
```{r}
Lamprey <- readRDS("../Species_Initial/Lamprey_initial.rds")
meta <- Lamprey@meta.data
counts <- Lamprey@assays$RNA@counts

unique_genes <- names(which(table(genes$Gene) == 1))
dup_genes <- setdiff(genes$Gene, unique_genes)

# Unique genes
unique_ID <- genes[genes$Gene %in% unique_genes, ]
rownames(unique_ID) <- unique_ID$ID

# Remove these genes, as the new gene name is more complicated
unique_ID <- unique_ID[setdiff(rownames(unique_ID), c("SLC6A9", "SLC32A1")),]

replace <- rownames(counts) %in% unique_ID$ID

rownames(counts)[replace] <- unique_ID[rownames(counts)[replace], "Gene"]

# Duplicated genes
dup_ID <- genes[genes$Gene %in% dup_genes, ]
rownames(dup_ID) <- dup_ID$ID

edit <- counts[1:length(dup_genes),]
rownames(edit) <- dup_genes
for(dup in dup_genes){
  IDs <- dup_ID[dup_ID$Gene == dup, "ID"]
  edit[dup,] <- colSums(counts[IDs, ])
}

counts <- rbind(counts, edit)

Lamprey <- CreateSeuratObject(counts)
Lamprey <- ClusterSeurat(Lamprey)
Lamprey@meta.data <- meta
Idents(Lamprey) <- "cell_class"

saveRDS(Lamprey, "../Species_Initial/Lamprey_initial.rds")
```



# Scratch
```{r}
Cow <- readRDS("../Species_Objects/SheepBC_integrated_v1.rds")
```



