---
title: "R Notebook"
output: html_notebook
---

## Load libraries

```{r}
library(tidyverse)
#library(ggtree)
# library(velocyto.R)
library(Seurat)
library(ggplot2)
library(cowplot)
source("../utils/utilFxns.R")
source("../utils/plottingFxns.R")
source("../utils/xgboost_train.R")
```

## Load velocyto data
Set data directory and load loom files
```{r}
data_dir <- "/Users/joshhahn/Google Drive File Stream/My Drive/shekharlab_data/projects/Peromyscus/Processed/190614_HY2HTCCXY/"
P1CD73S1 <- read.loom.matrices(file = paste0(data_dir, "Pero1CD73dpS1/velo_outs/possorted_genome_bam_C50GS.loom"), engine = "hdf5r")
P1CD73S2 <- read.loom.matrices(file = paste0(data_dir, "Pero1CD73dpS2/velo_outs/possorted_genome_bam_TR0D2.loom"), engine = "hdf5r")
P1CD73S3 <- read.loom.matrices(file = paste0(data_dir, "Pero1CD73dpS3/velo_outs/possorted_genome_bam_8FK0T.loom"), engine = "hdf5r")
P1CD90S1 <- read.loom.matrices(file = paste0(data_dir, "Pero1CD90S1/velo_outs/possorted_genome_bam_T2EZA.loom"), engine = "hdf5r")
P1CD90S2 <- read.loom.matrices(file = paste0(data_dir, "Pero1CD90S2/velo_outs/possorted_genome_bam_5USJR.loom"), engine = "hdf5r")
P2CD73S1 <- read.loom.matrices(file = paste0(data_dir, "Pero2CD73dpS1/velo_outs/possorted_genome_bam_J7I9L.loom"), engine = "hdf5r")
P2CD73S2 <- read.loom.matrices(file = paste0(data_dir, "Pero2CD73dpS2/velo_outs/possorted_genome_bam_WQV6I.loom"), engine = "hdf5r")
P3CD73S2 <- read.loom.matrices(file = paste0(data_dir, "Pero3CD73dpS2/velo_outs/possorted_genome_bam_KXVGF.loom"), engine = "hdf5r")
```

Add introns and exons and combine matrices
```{r}
P1CD73S1_mat = P1CD73S1$spliced + P1CD73S1$unspliced
P1CD73S2_mat = P1CD73S2$spliced + P1CD73S2$unspliced
P1CD73S3_mat = P1CD73S3$spliced + P1CD73S3$unspliced

P1CD90S1_mat = P1CD90S1$spliced + P1CD90S1$unspliced
P1CD90S2_mat = P1CD90S2$spliced + P1CD90S2$unspliced

P2CD73S1_mat = P2CD73S1$spliced + P2CD73S1$unspliced
P2CD73S2_mat = P2CD73S2$spliced + P2CD73S2$unspliced

P3CD73S2_mat = P3CD73S2$spliced + P3CD73S2$unspliced

Pero_mat <- cbind(P1CD73S1_mat, P1CD73S2_mat, P1CD73S3_mat, P1CD90S1_mat,P1CD90S2_mat, P2CD73S1_mat, P2CD73S2_mat, P3CD73S2_mat)
```

Create Seurat object and set file information
```{r}
Peromyscus <- CreateSeuratObject(Pero_mat, names.delim = ":")
Peromyscus@meta.data[colnames(P1CD73S1_mat), 'orig.file'] = "Pero1CD73dpS1"
Peromyscus@meta.data[colnames(P1CD73S2_mat), 'orig.file'] = "Pero1CD73dpS2"
Peromyscus@meta.data[colnames(P1CD73S3_mat), 'orig.file'] = "Pero1CD73dpS3"
Peromyscus@meta.data[colnames(P1CD90S1_mat), 'orig.file'] = "Pero1CD90S1"
Peromyscus@meta.data[colnames(P1CD90S2_mat), 'orig.file'] = "Pero1CD90S2"
Peromyscus@meta.data[colnames(P2CD73S1_mat), 'orig.file'] = "Pero2CD73dpS1"
Peromyscus@meta.data[colnames(P2CD73S2_mat), 'orig.file'] = "Pero2CD73dpS2"
Peromyscus@meta.data[colnames(P3CD73S2_mat), 'orig.file'] = "Pero3CD73dpS2"

Peromyscus@meta.data[colnames(P1CD73S1_mat), 'animal'] = 1
Peromyscus@meta.data[colnames(P1CD73S2_mat), 'animal'] = 1
Peromyscus@meta.data[colnames(P1CD73S3_mat), 'animal'] = 1
Peromyscus@meta.data[colnames(P1CD90S1_mat), 'animal'] = 1
Peromyscus@meta.data[colnames(P1CD90S2_mat), 'animal'] = 1
Peromyscus@meta.data[colnames(P2CD73S1_mat), 'animal'] = 2
Peromyscus@meta.data[colnames(P2CD73S2_mat), 'animal'] = 2
Peromyscus@meta.data[colnames(P3CD73S2_mat), 'animal'] = 3
```

## Cell class annotation
Cluster using general pipeline
```{r}
Peromyscus <- ClusterSeurat(Peromyscus)
saveRDS(Peromyscus, file = "../Species_Objects/Peromyscus_initial.rds")
```

Visualize initial clusters
```{r}
Peromyscus <- readRDS("../Species_Objects/Peromyscus_initial.rds")
DimPlot(Peromyscus, label = TRUE)
DimPlot(Peromyscus, group.by = "orig.file", cells = sample(colnames(Peromyscus)))
VlnPlot(Peromyscus, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Peromyscus, "nFeature_RNA", pt.size = 0) + RotatedAxis()
```

Annotate clusters using DE genes and Dot plot of marker genes
```{r}
Peromyscus_genes <- FindAllMarkers(Peromyscus)
saveRDS(Peromyscus_genes, file = "../Marker_Genes/Peromyscus_genes_initial.rds")
```

Look for major retinal cell classes
```{r}
# Peromyscus <- UpperCase_genes(Peromyscus)
Peromyscus <- DendroOrder(Peromyscus, nfeatures = 2000)
RGC_markers= c("RBPMS", "SLC17A6")
BC_markers=c("VSX2", "OTX2")
AC_markers = c("TFAP2A", "GAD1", "GAD2", "SLC6A9")
HC_markers=c("ONECUT1","ONECUT2")
Cone_markers = c("ARR3", "OPN1SW", "OPN1MW")
Rod_markers = c("RHO", "PDC")
MG_markers = c("GLUL", "APOE", "RLBP1")

DotPlot(Peromyscus, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Rod_markers,Cone_markers, MG_markers), group.by = "dendro_order") + RotatedAxis()
```


Annotate clusters - USE INTEGRATED IDENTITIES
```{r}
Peromyscus@meta.data$cell_class = "Other"
Peromyscus@meta.data[WhichCells(Peromyscus, idents =c(48,2,7,11,35,9,39,16,51,20,15,18)),]$cell_class = "RGC"
Peromyscus@meta.data[WhichCells(Peromyscus, idents = c()),]$cell_class = "BP"
Peromyscus@meta.data[WhichCells(Peromyscus, idents = c(6,52,31,32,21,44,49,3,33,24,34,22,10,27,4,54,37,53)),]$cell_class = "AC"
Peromyscus@meta.data[WhichCells(Peromyscus, idents = c(45)),]$cell_class = "HC"
Peromyscus@meta.data[WhichCells(Peromyscus, idents = c(41)),]$cell_class = "Rod"
Peromyscus@meta.data[WhichCells(Peromyscus, idents = c(42)),]$cell_class = "Cone"

DimPlot(Peromyscus, group.by = "cell_class")

# Questionable: 14, 26
markers14 <- FindMarkers(Peromyscus, ident.1 = 14)

saveRDS(Peromyscus, "../Species_Objects/Peromyscus_initial.rds")
Peromyscus <- readRDS("../Species_Objects/Peromyscus_initial.rds")
```

Plot cell type by file
```{r}
counts <- table(Peromyscus@meta.data$cell_class, Peromyscus@meta.data$orig.file)
counts
counts <- t(t(counts) / colSums(counts))
barplot(counts, legend = rownames(counts), col= c("red","orange","yellow","green","blue", "black"))
```

Perform batch integration
```{r}
library(Seurat)
Peromyscus <- readRDS("../Species_Initial/Peromyscus_initial.rds")
# Split object
object.list <- SplitObject(Peromyscus, split.by = "animal")
rm(Peromyscus)
# Normalize each dataset and find variable features
for (i in 1:length(object.list)) {
  object.list[[i]] <- NormalizeData(object.list[[i]])
  object.list[[i]] <- FindVariableFeatures(object.list[[i]], selection.method = "vst", 
                                            nfeatures = 1500)
  }
# Find Integration anchors
object.anchors <- FindIntegrationAnchors(object.list = object.list, dims = 1:40)
saveRDS(object.anchors, "../Species_Objects/Pero_anchors.rds")

# Integrate Data
Peromyscus.i <- IntegrateData(anchorset = object.anchors, dims = 1:40)
saveRDS(Peromyscus.i, "../Species_Objects/Peromysucs_integrated_initial_0105.rds")
```

Repeat clustering using integrated object
```{r}
Peromyscus.i <- readRDS("../Species_Objects/Peromyscus_integrated_initial_0105.rds")
DefaultAssay(Peromyscus.i) <- "integrated"
# Run the standard workflow for visualization and clustering
Peromyscus.i <- ClusterSeurat(Peromyscus.i, integrated = TRUE)
saveRDS(Peromyscus.i, file = "../Species_Objects/Peromyscus_integrated_initial_0105.rds")
```

Compare integrated versus non integrated assays
```{r}
# Peromyscus <- readRDS("../Species_Objects/Peromyscus_initial.rds")
DimPlot(Peromyscus.i)
DimPlot(Peromyscus.i, reduction = "umap", group.by = "orig.file") +ggtitle("Integrated")
DimPlot(Peromyscus, reduction = "umap", group.by = "orig.file") + ggtitle("Non-integrated")
```


Annotate clusters using DE genes and Dot plot of marker genes
```{r}
Idents(Peromyscus.i) <- "integrated_snn_res.0.5"
Peromyscus.i_genes <- FindAllMarkers(Peromyscus.i)
saveRDS(Peromyscus.i_genes, file = "../Marker_Genes/Peromyscus_integrated_genes_initial.rds")
```

Check again for major retinal cell classes
```{r}
Peromyscus.i <- readRDS("../Species_Objects/Peromyscus_integrated_initial_0105.rds")
Peromyscus.i <- UpperCase_genes(Peromyscus.i)
Peromyscus.i <- DendroOrder(Peromyscus.i)

RGC_markers= c("RBPMS", "SLC17A6")
BC_markers=c("VSX2", "OTX2")
AC_markers = c("TFAP2A", "GAD1", "GAD2", "SLC6A9")
HC_markers=c("ONECUT1","ONECUT2")
Cone_markers = c("ARR3", "OPN1SW", "OPN1MW")
Rod_markers = c("RHO", "PDC")
MG_markers = c("GLUL", "APOE", "RLBP1")

DotPlot(Peromyscus.i, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Cone_markers,Rod_markers, MG_markers), group.by = "dendro_order", assay = "RNA") + RotatedAxis()
```

Annotate clusters
```{r}
Peromyscus.i <- readRDS("../Species_Objects/Peromyscus_integrated_initial_0105.rds")
Idents(Peromyscus.i) <- "integrated_snn_res.0.5"

Peromyscus.i@meta.data$int_cell_class = "Other"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents =c(25,40,45,50,23,6,3,5,12,16,22,24)),]$int_cell_class = "RGC"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(9,1,4,21, 39, 14)),]$int_cell_class = "BP"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(36,17,33,46,11,47,31,44,28,7,15,2,30,27,18,34)),]$int_cell_class = "GabaAC"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(19,8,32, 20,10,13,37,29)),]$int_cell_class = "GlyAC"

Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(43)),]$int_cell_class = "HC"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(38)),]$int_cell_class = "Rod"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(42)),]$int_cell_class = "Cone"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(0,35,26,41)),]$int_cell_class = "MG"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(49)),]$int_cell_class = "MicroG"

Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(48)),]$int_cell_class = "Other"


DimPlot(Peromyscus.i, group.by = "int_cell_class")
saveRDS(Peromyscus.i, "../Species_Objects/Peromyscus_integrated_initial_0105.rds")
```

Save as initial object, change default assay to RNA
```{r}
DefaultAssay(Peromyscus.i) <- "RNA"
Peromyscus.i@meta.data$cell_class <- Peromyscus.i@meta.data$int_cell_class
saveRDS(Peromyscus.i, "../Species_Objects/Peromyscus_initial.rds")
```


```{r}
VlnPlot(Peromyscus.i, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Peromyscus.i, "nFeature_RNA", pt.size = 0) + RotatedAxis()
```


Investigate clusters not annotated as above
```{r}
Pero_genes <- readRDS("../Marker_Genes/Peromyscus_integrated_genes_initial.rds")

# Cluster 10 - contains SLC6A9, amacrine 
cat("Cluster 10\n")
head(subset(Pero_genes, cluster == 10), 10)$gene
# Cluster 13 - contains SLC6A9, amacrine 
cat("\nCluster 13\n")
head(subset(Pero_genes, cluster == 13), 10)$gene
# Cluster 37 - contains SLC6A9, amacrine
cat("\nCluster 37\n")
head(subset(Pero_genes, cluster == 37), 10)$gene

# Cluster 14 - annotate bipolar, contains GRIK2
cat("\nCluster 14\n")
head(subset(Pero_genes, cluster == 14), 10)$gene

# Cluster 29 - contains SLC6A9, amacrine
cat("\nCluster 29\n")
head(subset(Pero_genes, cluster == 29), 10)$gene

# Cluster 48 - annotate as RGC since within RGC branch on dendrogram
cat("\nCluster 48\n")
head(subset(Pero_genes, cluster == 48), 10)$gene

# Cluster 49 - microglia
cat("\nCluster 49\n")
head(subset(Pero_genes, cluster == 49), 10)$gene

Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents =c(48)),]$int_cell_class = "RGC"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(14)),]$int_cell_class = "BC"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(10,13,37,29)),]$int_cell_class = "AC"
Peromyscus.i@meta.data[WhichCells(Peromyscus.i, idents = c(49)),]$int_cell_class = "MG"

DimPlot(Peromyscus.i, group.by = "int_cell_class")
saveRDS(Peromyscus.i, "../Species_Objects/Peromyscus_integrated_initial_0105.rds")
```


```{r}
Peromyscus.i <- readRDS("../Species_Objects/Peromyscus_integrated_initial_0105.rds")
counts <- table(Peromyscus.i@meta.data$int_cell_class, Peromyscus.i@meta.data$orig.file)
counts
counts <- t(t(counts) / colSums(counts))
barplot(counts, legend = rownames(counts), col= rev(c("red","orange","yellow","white","green","blue")))
```

## RGCs
Create initial object
```{r}
Pero <- readRDS("../Species_Objects/Peromyscus_integrated_initial_0105.rds")

# Only keep RGCs
Idents(Pero) <- "int_cell_class"
Pero_RGC <- subset(Pero, cells = WhichCells(Pero, idents ="RGC"))
rm(Pero)

# Only keep cells from samples that enrich for RGCs
Idents(Pero_RGC) <- "orig.file"
Pero_RGC <- subset(Pero_RGC, cells = WhichCells(Pero_RGC, idents =c("Pero1CD90S1", "Pero1CD90S2")))

VlnPlot(Pero_RGC, "nCount_RNA", group.by = "orig.file", pt.size = 0)
VlnPlot(Pero_RGC, "nFeature_RNA", group.by = "orig.file", pt.size = 0)

Pero_RGC <- subset(Pero_RGC, nFeature_RNA > 50 & nCount_RNA < 30000)

saveRDS(Pero_RGC, file = "../Species_Objects/PeromyscusRGC_v1.rds")
```


Cluster RGCs, batch correct by file
```{r}
# Perform integration - integrate by file
obj.list <- SplitObject(Pero_RGC, split.by = "orig.file")
for (i in 1:length(obj.list)) {
      obj.list[[i]] <- NormalizeData(obj.list[[i]], verbose = FALSE)
      obj.list[[i]] <- FindVariableFeatures(obj.list[[i]], selection.method = "vst", nfeatures = 2000, verbose = FALSE)
}
obj.anchors <- FindIntegrationAnchors(object.list = obj.list)
Pero_RGC <- IntegrateData(anchorset = obj.anchors)
DefaultAssay(Pero_RGC) <- "integrated"

Pero_RGC <- ScaleData(Pero_RGC, verbose = FALSE)
Pero_RGC <- RunPCA(Pero_RGC, npcs = 30, verbose = FALSE)
ElbowPlot(Pero_RGC, ndims = 30)  

# 25 PCs seems to suffciently capture the data
nPCs <- 25
Pero_RGC <- FindNeighbors(Pero_RGC, dims = 1:nPCs)
# Start with a resolution parameter of 1.
Pero_RGC <- FindClusters(Pero_RGC, resolution = 1)
Pero_RGC <- RunTSNE(Pero_RGC, dims = 1:nPCs)
Pero_RGC <- RunUMAP(Pero_RGC, dims = 1:nPCs)

saveRDS(Pero_RGC, "~/Species_Objects/PeromyscusRGC_integrated_v1.rds")

```

Visualize initial clusters
```{r}
# Pero_RGC <- readRDS("~/Species_Objects/PeromyscusRGC_integrated_v1.rds")
Idents(Pero_RGC) <- "seurat_clusters"
DimPlot(Pero_RGC, label = TRUE, group.by = "integrated_snn_res.1")
DimPlot(Pero_RGC, group.by = "orig.file", cells = sample(colnames(Pero_RGC)))
VlnPlot(Pero_RGC, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Pero_RGC, "nFeature_RNA", pt.size = 0) + RotatedAxis()
DotPlot(Pero_RGC, features = c("RBPMS", "SLC17A6", "THY1"), assay = "RNA")
```

Calculate DE markers using the Wilcoxon test to determine if clusters need to be merged
```{r}
Idents(Pero_RGC) <- "seurat_clusters"
PeroRGC_markers <- FindAllMarkers(Pero_RGC, assay = "RNA", only.pos = TRUE)
saveRDS(PeroRGC_markers, "../Species_Markers/PeroRGCMarkers_v1.rds")
```

Plot top markers for each cluster, arranged by a dendrogram
```{r}
#PeroRGC_markers <- readRDS("../Species_Markers/PeroRGCMarkers_v1.rds")
#Pero_RGC <- readRDS("../Species_Objects/PeromyscusRGC_v1.rds")
Pero_RGC <- DendroOrder(Pero_RGC)

Idents(Pero_RGC) <- "dendro_order"
top_markers <- TopMarkers(Pero_RGC, markers = PeroRGC_markers, num_markers = 2)

DotPlot(Pero_RGC, features = top_markers, assay = "RNA", group.by = "dendro_order") + RotatedAxis()
```

Merge clusters based on lack of DE genes
```{r}
# 0, 27: 27 retains unique markers, keep separate for now
DotPlot(Pero_RGC, features = head(rownames(FindMarkers(Pero_RGC, ident.1 = 0, ident.2 = 27)),10), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

# 3, 15: some gene signatures are different
DotPlot(Pero_RGC, features = head(rownames(FindMarkers(Pero_RGC, ident.1 = 3, ident.2 = 15)),10), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

# Remove cluster 4: low counts, no DE genes
Pero_RGC <- DropClusters(Pero_RGC, idents = 4, refactor = FALSE)

# Cluster 11: other genes are uniquely expressed in other clusters, but combo is unique
DotPlot(Pero_RGC, features = head(subset(PeroRGC_markers, cluster == 11),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()

# 32 and 10: 32 has several unique genes, keep separate for now
DotPlot(Pero_RGC, features = head(subset(PeroRGC_markers, cluster == 32),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()
DotPlot(Pero_RGC, features = head(rownames(FindMarkers(Pero_RGC, ident.1 = 32, ident.2 = 10)),10), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

# 6 and 24: merge, no unique genes between the two
DotPlot(Pero_RGC, features = head(subset(PeroRGC_markers, cluster == 6),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()
DotPlot(Pero_RGC, features = head(rownames(FindMarkers(Pero_RGC, ident.1 = 6, ident.2 = 24)),10), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

Pero_RGC <- MergeClusters(Pero_RGC, idents = c(6,24), refactor = TRUE)

# 1 and 5: some unique genes, keep separate for now
DotPlot(Pero_RGC, features = head(subset(PeroRGC_markers, cluster == 1),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()
DotPlot(Pero_RGC, features = head(rownames(FindMarkers(Pero_RGC, ident.1 = 1, ident.2 = 5)),10), group.by = "dendro_order", assay = "RNA") + RotatedAxis()


saveRDS(Pero_RGC, file = "../Species_Objects/PeromyscusRGC_integrated_v2.rds")
```

Recalculate DE genes 
```{r}
Pero_RGC <- readRDS("../Species_Objects/PeromyscusRGC_integrated_v2.rds")
Idents(Pero_RGC) <- "seurat_clusters"
PeroRGC_markers <- FindAllMarkers(Pero_RGC, assay = "RNA", test.use = "MAST")
saveRDS(PeroRGC_markers, "../Species_Markers/PeroRGCMarkers_v2.rds")
```



### Check for doublets
Convert UMAP and diagonal gene plots to files
```{r}
Pero_RGC <- readRDS("../Species_Objects/PeromyscusRGC_integrated_v2.rds")

RGC_markers= c("RBPMS", "SLC17A6")
BC_markers=c("VSX2", "OTX2")
AC_markers = c("TFAP2A", "GAD1", "GAD2", "SLC6A9")
HC_markers=c("ONECUT1","ONECUT2")
Cone_markers = c("ARR3", "OPN1SW", "OPN1MW")
Rod_markers = c("RHO", "PDC")
MG_markers = c("GLUL", "APOE", "RLBP1")

DotPlot(Pero_RGC, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Rod_markers,Cone_markers, MG_markers), assay="RNA") + RotatedAxis()

# Cluster 25 is AC doublet
Pero_RGC <- DropClusters(Pero_RGC, idents = 25, refactor = TRUE)

Pero_RGC@meta.data$type <- Pero_RGC@meta.data$seurat_clusters
Idents(Pero_RGC) <- "type"

saveRDS(Pero_RGC, "../Species_Objects/PeromyscusRGC_integrated_v3.rds")
Pero_RGC <- readRDS("../Species_Objects/PeromyscusRGC_integrated_v3.rds")


```


### Convert UMAP and diagonal gene plots to files
```{r}
Pero_RGC <- readRDS("../Species_Objects/PeromyscusRGC_integrated_v3.rds")
BC_markers <- readRDS()


pdf("../Figures/Species_UMAPs/PeromyscusRGC.pdf", w=4, h=4, useDingbats = FALSE)
DimPlot(Pero_RGC, label = TRUE, raster = TRUE) + NoLegend()
dev.off()

markers <- PlotUniqueMarkers(Pero_RGC, RGC_markers, edits = TRUE)
pdf("Figures/Species_MarkerPlots/PeromyscusRGC.pdf", , w=6, h=4, useDingbats = FALSE)
dev.off()

```
## BCs
Create initial object
```{r}
Pero <- readRDS("../Species_Objects/Peromyscus_integrated_initial_0105.rds")

# Only keep BCs
Idents(Pero) <- "int_cell_class"
Pero_BC <- subset(Pero, cells = WhichCells(Pero, idents ="BC"))
rm(Pero)

# Remove cells from samples that do not enrich for BCs
Idents(Pero_BC) <- "orig.file"
Pero_BC <- subset(Pero_BC, cells = setdiff(colnames(Pero_BC), WhichCells(Pero_BC, idents =c("Pero1CD90S1", "Pero1CD90S2"))))

VlnPlot(Pero_BC, "nCount_RNA", group.by = "orig.file", pt.size = 0)
VlnPlot(Pero_BC, "nFeature_RNA", group.by = "orig.file", pt.size = 0)

Pero_BC <- subset(Pero_BC, nFeature_RNA > 50 & nCount_RNA < 10000)

saveRDS(Pero_BC, file = "../Species_Objects/PeromyscusBC_v1.rds")
```


### Cluster BCs, batch correct by file
```{r}
Pero_BC <- readRDS("../Species_Objects/PeromyscusBC_v1.rds")
# Perform integration - integrate by file
obj.list <- SplitObject(Pero_BC, split.by = "orig.file")
for (i in 1:length(obj.list)) {
      obj.list[[i]] <- NormalizeData(obj.list[[i]], verbose = FALSE)
      obj.list[[i]] <- FindVariableFeatures(obj.list[[i]], selection.method = "vst", nfeatures = 2000, verbose = FALSE)
}
obj.anchors <- FindIntegrationAnchors(object.list = obj.list)
Pero_BC <- IntegrateData(anchorset = obj.anchors)
DefaultAssay(Pero_BC) <- "integrated"

Pero_BC <- ScaleData(Pero_BC, verbose = FALSE)
Pero_BC <- RunPCA(Pero_BC, npcs = 30, verbose = FALSE)
ElbowPlot(Pero_BC, ndims = 30)  

# 20 PCs seems to suffciently capture the data
nPCs <- 20
Pero_BC <- FindNeighbors(Pero_BC, dims = 1:nPCs)
# Start with a resolution parameter of 1.
# Parameter seems a bit high: reduce to .8
Pero_BC <- FindClusters(Pero_BC, resolution = .8)
Pero_BC <- RunTSNE(Pero_BC, dims = 1:nPCs)
Pero_BC <- RunUMAP(Pero_BC, dims = 1:nPCs)

saveRDS(Pero_BC, "~/Species_Objects/PeromyscusBC_integrated_v1.rds")

```

### Visualize initial clusters
```{r}
# Pero_BC <- readRDS("~/Species_Objects/PeromyscusBC_integrated_v1.rds")
DimPlot(Pero_BC, label = TRUE, group.by = "integrated_snn_res.1")
DimPlot(Pero_BC, group.by = "orig.file", cells = sample(colnames(Pero_BC)))
VlnPlot(Pero_BC, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Pero_BC, "nFeature_RNA", pt.size = 0) + RotatedAxis()
```

### Calculate DE genes for contaminant removal and merging spurious clusters
```{r}
PeroBC_markers <- FindAllMarkers(Pero_BC, assay = "RNA", only.pos = TRUE)
saveRDS(PeroBC_markers, "../Species_Markers/PeroBCMarkers_v1.rds")
```

Plot top markers for each cluster, arranged by a dendrogram
```{r}
# PeroBC_markers <- readRDS("../Species_Markers/PeroBCMarkers_v1.rds")
# Pero_BC <- readRDS("../Species_Objects/PeromyscusBC_v1.rds")
Idents(Pero_BC) <- "integrated_snn_res.1"
Pero_BC <- DendroOrder(Pero_BC)

Idents(Pero_BC) <- "dendro_order"
top_markers <- TopMarkers(Pero_BC, markers = PeroBC_markers, num_markers = 3)

DotPlot(Pero_BC, features = top_markers, assay = "RNA", group.by = "dendro_order") + RotatedAxis()
```

### Merge clusters based on lack of DE genes
```{r}
# 7, 20: 20 has unique markers, keep separate for now
DotPlot(Pero_BC, features = head(rownames(FindMarkers(Pero_RGC, ident.1 = 7, ident.2 = 20)),10), group.by = "dendro_order", assay = "RNA") + RotatedAxis()
DotPlot(Pero_BC, features = head(subset(PeroBC_markers, cluster == 7),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()

# 16, 0, 3, 1 - merge
DotPlot(Pero_BC, features = head(subset(PeroBC_markers, cluster == 16),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()
DotPlot(Pero_BC, features = head(rownames(FindMarkers(Pero_BC, ident.1 = 0, ident.2 = c(3,16)),10)), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

Pero_BC <- MergeClusters(Pero_BC, idents = c(0,1,3,16), refactor = FALSE)

# 17 - Photoreceptors, remove
DotPlot(Pero_BC, features = head(subset(PeroBC_markers, cluster == 17),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()

Pero_BC <- DropClusters(Pero_BC, idents = 17, refactor = FALSE)

# 18, 5 - merge
DotPlot(Pero_BC, features = head(subset(PeroBC_markers, cluster == 18),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()
DotPlot(Pero_BC, features = head(rownames(FindMarkers(Pero_BC, ident.1 = 18, ident.2 = 5),10)), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

Pero_BC <- MergeClusters(Pero_BC, idents = c(5,18), refactor = FALSE)

# 13, 4 - merge
# 11- keep separate 
DotPlot(Pero_BC, features = head(subset(PeroBC_markers, cluster == 13),10)$gene, group.by = "dendro_order", assay = "RNA") + RotatedAxis()
DotPlot(Pero_BC, features = head(rownames(FindMarkers(Pero_BC, ident.1 = 13, ident.2 = 4),10)), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

Pero_BC <- MergeClusters(Pero_BC, idents = c(4,13), refactor = TRUE)

saveRDS(Pero_BC, file = "../Species_Objects/PeromyscusBC_integrated_v2.rds")
```



### Look for doublets
```{r}
Pero_BC <- readRDS("../Species_Objects/PeromyscusBC_integrated_v2.rds")

# Clusters 3 and 10 potentially AC doublets
BC_markers=c("VSX2", "CABP5", "OTX2", "PRKCA", "TMEM215")
AC_markers=c("TFAP2A","TFAP2B", "GAD1", "C1QL1", "C1QL2", "SLC6A9", "PAX6")
HC_markers=c("RET", "ONECUT1","ONECUT2", "CALB1", "TPM3", "NRSN1")
PR_markers=c("PDE6H", "CRX", "SAG", "PDC", "RHO", "ARR3")
MG_markers=c("SLC1A3","RLBP1", "GLUL", "APOE")
Other_markers = c("FN1", "GSN","CLDN5","RGS5")

DotPlot(Pero_BC, features = c(BC_markers, AC_markers, HC_markers, PR_markers, MG_markers, Other_markers), assay = "RNA") + RotatedAxis()
```

```{r}
Pero_BC <- DropClusters(Pero_BC, idents = c(3,10), refactor = TRUE)
Pero_BC@meta.data$type <- Pero_BC@meta.data$seurat_clusters
Idents(Pero_BC) <- "type"
DefaultAssay(Pero_BC) <- "RNA"

DefaultAssay(Pero_BC) <- "integrated"

# Start with a resolution parameter of 1.
# Parameter seems a bit high: reduce to .8
Pero_BC <- FindNeighbors(Pero_BC, dims = 1:20)
Pero_BC <- FindClusters(Pero_BC, resolution = .6)
Pero_BC <- RunUMAP(Pero_BC, dims = 1:20)

saveRDS(Pero_BC, "../Species_Objects/PeromyscusBC_integrated_v3.rds")
```






### Recalculate DE genes 
```{r}
Pero_BC <- readRDS("../Species_Objects/PeromyscusBC_integrated_v2.rds")
Idents(Pero_BC) <- "seurat_clusters"
PeroBC_markers <- FindAllMarkers(Pero_BC, assay = "RNA", test.use = "MAST")
saveRDS(PeroBC_markers, "../Species_Markers/PeroBCMarkers_v2.rds")
```

### Convert UMAP and diagonal gene plots to files
```{r}
Pero_BC <- readRDS("../Species_Objects/PeromyscusBC_integrated_v3.rds")


pdf("../Figures/Species_UMAPs/PeromyscusBC.pdf", w=4, h=4, useDingbats = FALSE)
DimPlot(Pero_BC, label = TRUE, raster = TRUE) + NoLegend()
dev.off()

markers <- PlotUniqueMarkers(Pero_RGC, RGC_markers, edits = TRUE)
pdf("Figures/Species_MarkerPlots/PeromyscusRGC.pdf", , w=6, h=4, useDingbats = FALSE)
dev.off()

```

### Mouse comparison
```{r}
# Read in orthologs and data
Mouse_BC <- readRDS("../Species_Reference/DropseqBC.rds")
Pero_BC <- readRDS("../Species_Objects/PeromyscusBC_integrated_v2.rds")
DefaultAssay(Pero_BC) <- "RNA"
orthologs <- readRDS(file = "../../orthology_table.rds")

# Only keep bipolars from Mouse object
Mouse_BC <- subset(Mouse_BC, cells = setdiff(colnames(Mouse_BC), WhichCells(Mouse_BC, idents = c("GabaAC", "MG", "GlyAC", "Rods", "Cones"))))

# Subset to genes that are present in each object
orthologs <- subset(orthologs, (toupper(orthologs$Mouse.gene.name) %in% rownames(DropseqBC))  & (toupper(orthologs$Peromyscus.gene.name) %in% rownames(Pero_BC)))

# Create a trained classifier on Mouse bipolar Dropseq data
MouseBC_model <- TrainModel(object = Mouse_BC, training_genes =toupper(orthologs$Mouse.gene.name))
```

Compare to Peromycus
```{r}
P_M_mat <- BuildConfusionMatrix(test = Pero_BC, train = Mouse_BC, model = MouseBC_model)
MakePrettyConfusionMatrix(P_M_mat, xlab.use = "Mouse Bipolar Clusters", ylab.use = "Peromyscus Bipolar Clusters")
```

```{r}
PeroBC_markers <- readRDS("../Species_Markers/")
```


## Rods
```{r}
Peromyscus <- readRDS("../Species_Objects/Peromyscus_integrated_initial_0105.rds")
Idents(Peromyscus) <- "integrated_snn_res.0.5"

Rod_markers <- c("RHO" ,"PDC", "NRL", "SAG", "GNAT1", "GNGT1")
Cone_markers <- c("ARR3", "PDE6H", "GNAT2","GNGT2","OPN1SW","OPN1MW")
DotPlot(Peromyscus, features = c(Rod_markers, Cone_markers), group.by = "dendro_order", assay = "RNA") + RotatedAxis()

# Subset to rods, do preliminary clustering, and save
Peromyscus_Rods <- subset(Peromyscus, cells = WhichCells(Peromyscus, idents = 38))
DefaultAssay(Peromyscus_Rods) <- "RNA"
Peromyscus_Rods <- ClusterSeurat(Peromyscus_Rods)

saveRDS(Peromyscus_Rods, file = "../Species_Objects/PeromyscusRods_v1.rds")
```
