---
title: "R Notebook"
output: html_notebook
---

# Load libraries
```{r}
library(tidyverse)
library(ggtree)
library(Seurat)
library(ggplot2)
source("../utils/utilFxns.R")
source("../utils/xgboost_train.R")
```


# Prepare velocyto data
```{r}
data_dir <- "/Users/joshhahn/Google Drive File Stream/My Drive/shekharlab_data/projects/Zebrafish/Loom/Bipolar/"
ZfishBC1 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC1/possorted_genome_bam_OQDY2.loom"), engine = "hdf5r")
ZfishBC2 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC2/possorted_genome_bam_TIPBO.loom"), engine = "hdf5r")
ZfishBC3 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC3/possorted_genome_bam_S06NP.loom"), engine = "hdf5r")
ZfishBC4 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC4/possorted_genome_bam_S557J.loom"), engine = "hdf5r")
ZfishBC5 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC5/possorted_genome_bam_QFA52.loom"), engine = "hdf5r")
ZfishBC6 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC6/possorted_genome_bam_AC7O9.loom"), engine = "hdf5r")
ZfishBC7 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC7/possorted_genome_bam_7JH2I.loom"), engine = "hdf5r")
ZfishBC8 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC8/possorted_genome_bam_MROMJ.loom"), engine = "hdf5r")

data_dir <- "/Users/joshhahn/Google Drive File Stream/My Drive/shekharlab_data/projects/Zebrafish/Loom/Full/"
ZfishRGC1 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC1/possorted_genome_bam_5FL41.loom"), engine = "hdf5r")
ZfishRGC2 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC2/possorted_genome_bam_55PX4.loom"), engine = "hdf5r")
ZfishRGC3 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC3/possorted_genome_bam_CKQGO.loom"), engine = "hdf5r")
ZfishRGC4 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC4/possorted_genome_bam_QNYFS.loom"), engine = "hdf5r")
ZfishRGC5 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC5/possorted_genome_bam_WJ3K3.loom"), engine = "hdf5r")
ZfishRGC6 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC6/possorted_genome_bam_WKX5P.loom"), engine = "hdf5r")
ZfishRGC7 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC7/possorted_genome_bam_RJ1QQ.loom"), engine = "hdf5r")
ZfishRGC8 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC8/possorted_genome_bam_B90PX.loom"), engine = "hdf5r")
ZfishRGC9 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC9/possorted_genome_bam_FSVZ6.loom"), engine = "hdf5r")
ZfishRGC10 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC10/possorted_genome_bam_I4XQT.loom"), engine = "hdf5r")
ZfishRGC11 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC11/possorted_genome_bam_7NZAN.loom"), engine = "hdf5r")
ZfishRGC12 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC12/possorted_genome_bam_BUEEB.loom"), engine = "hdf5r")
ZfishRGC13 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC13/possorted_genome_bam_SZB0L.loom"), engine = "hdf5r")
ZfishRGC14 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC14/possorted_genome_bam_HPP9K.loom"), engine = "hdf5r")
ZfishRGC15 <- read.loom.matrices(file = paste0(data_dir, "ZebrafishRGC15/possorted_genome_bam_ITV9O.loom"), engine = "hdf5r")
```

Add introns and exons and combine matrices
```{r}
ZfishBC1_mat <- ZfishBC1$spliced + ZfishBC1$unspliced
ZfishBC2_mat <- ZfishBC2$spliced + ZfishBC2$unspliced
ZfishBC3_mat <- ZfishBC3$spliced + ZfishBC3$unspliced
ZfishBC4_mat <- ZfishBC4$spliced + ZfishBC4$unspliced
ZfishBC5_mat <- ZfishBC5$spliced + ZfishBC5$unspliced
ZfishBC6_mat <- ZfishBC6$spliced + ZfishBC6$unspliced
ZfishBC7_mat <- ZfishBC7$spliced + ZfishBC7$unspliced
ZfishBC8_mat <- ZfishBC8$spliced + ZfishBC8$unspliced

ZfishRGC1_mat <- ZfishRGC1$spliced + ZfishRGC1$unspliced
ZfishRGC2_mat <- ZfishRGC2$spliced + ZfishRGC2$unspliced
ZfishRGC3_mat <- ZfishRGC3$spliced + ZfishRGC3$unspliced
ZfishRGC4_mat <- ZfishRGC4$spliced + ZfishRGC4$unspliced
ZfishRGC5_mat <- ZfishRGC5$spliced + ZfishRGC5$unspliced
ZfishRGC6_mat <- ZfishRGC6$spliced + ZfishRGC6$unspliced
ZfishRGC7_mat <- ZfishRGC7$spliced + ZfishRGC7$unspliced
ZfishRGC8_mat <- ZfishRGC8$spliced + ZfishRGC8$unspliced
ZfishRGC9_mat <- ZfishRGC9$spliced + ZfishRGC9$unspliced
ZfishRGC10_mat <- ZfishRGC10$spliced + ZfishRGC10$unspliced
ZfishRGC11_mat <- ZfishRGC11$spliced + ZfishRGC11$unspliced
ZfishRGC12_mat <- ZfishRGC12$spliced + ZfishRGC12$unspliced
ZfishRGC13_mat <- ZfishRGC13$spliced + ZfishRGC13$unspliced
ZfishRGC14_mat <- ZfishRGC14$spliced + ZfishRGC14$unspliced
ZfishRGC15_mat <- ZfishRGC15$spliced + ZfishRGC15$unspliced


Zebrafish_mat <- cbind(ZfishBC1_mat, ZfishBC2_mat, ZfishBC3_mat, ZfishBC4_mat, ZfishBC5_mat, ZfishBC6_mat, ZfishBC7_mat, ZfishBC8_mat, ZfishRGC1_mat, ZfishRGC2_mat, ZfishRGC3_mat, ZfishRGC4_mat, ZfishRGC5_mat, ZfishRGC6_mat, ZfishRGC7_mat, ZfishRGC8_mat, ZfishRGC9_mat, ZfishRGC10_mat, ZfishRGC11_mat, ZfishRGC12_mat, ZfishRGC13_mat, ZfishRGC14_mat, ZfishRGC15_mat)
```

Create Seurat object
```{r}
Zebrafish <- CreateSeuratObject(Zebrafish_mat, names.delim = ":")

Zebrafish@meta.data[colnames(ZfishBC1_mat), 'orig.file'] = "ZfishBC1"
Zebrafish@meta.data[colnames(ZfishBC2_mat), 'orig.file'] = "ZfishBC2"
Zebrafish@meta.data[colnames(ZfishBC3_mat), 'orig.file'] = "ZfishBC3"
Zebrafish@meta.data[colnames(ZfishBC4_mat), 'orig.file'] = "ZfishBC4"
Zebrafish@meta.data[colnames(ZfishBC5_mat), 'orig.file'] = "ZfishBC5"
Zebrafish@meta.data[colnames(ZfishBC6_mat), 'orig.file'] = "ZfishBC6"
Zebrafish@meta.data[colnames(ZfishBC7_mat), 'orig.file'] = "ZfishBC7"
Zebrafish@meta.data[colnames(ZfishBC8_mat), 'orig.file'] = "ZfishBC8"

Zebrafish@meta.data[colnames(ZfishRGC1_mat), 'orig.file'] = "ZfishRGC1"
Zebrafish@meta.data[colnames(ZfishRGC2_mat), 'orig.file'] = "ZfishRGC2"
Zebrafish@meta.data[colnames(ZfishRGC3_mat), 'orig.file'] = "ZfishRGC3"
Zebrafish@meta.data[colnames(ZfishRGC4_mat), 'orig.file'] = "ZfishRGC4"
Zebrafish@meta.data[colnames(ZfishRGC5_mat), 'orig.file'] = "ZfishRGC5"
Zebrafish@meta.data[colnames(ZfishRGC6_mat), 'orig.file'] = "ZfishRGC6"
Zebrafish@meta.data[colnames(ZfishRGC7_mat), 'orig.file'] = "ZfishRGC7"
Zebrafish@meta.data[colnames(ZfishRGC8_mat), 'orig.file'] = "ZfishRGC8"
Zebrafish@meta.data[colnames(ZfishRGC9_mat), 'orig.file'] = "ZfishRGC9"
Zebrafish@meta.data[colnames(ZfishRGC10_mat), 'orig.file'] = "ZfishRGC10"
Zebrafish@meta.data[colnames(ZfishRGC11_mat), 'orig.file'] = "ZfishRGC11"
Zebrafish@meta.data[colnames(ZfishRGC12_mat), 'orig.file'] = "ZfishRGC12"
Zebrafish@meta.data[colnames(ZfishRGC13_mat), 'orig.file'] = "ZfishRGC13"
Zebrafish@meta.data[colnames(ZfishRGC14_mat), 'orig.file'] = "ZfishRGC14"
Zebrafish@meta.data[colnames(ZfishRGC15_mat), 'orig.file'] = "ZfishRGC15"
```

Initial quality control metrics
```{r}
VlnPlot(Zebrafish, features = "nCount_RNA", pt.size = 0, group.by = "orig.file")
VlnPlot(Zebrafish, features = "nFeature_RNA", pt.size = 0, group.by = "orig.file")
```

# Cell class annotation

Perform initial clustering
```{r}
Zebrafish <- ClusterSeurat(Zebrafish)
Zebrafish <- NormalizeData(Zebrafish)
Zebrafish <- FindVariableFeatures(Zebrafish)
Zebrafish <- ScaleData(Zebrafish)
Zebrafish <- RunPCA(Zebrafish)
Zebrafish <- FindNeighbors(Zebrafish, dims = 1:30)
Zebrafish <- FindClusters(Zebrafish, resolution = .5)
Zebrafish <- RunUMAP(Zebrafish, dims = 1:30)
saveRDS(Zebrafish, file = "../Species_Objects/Zebrafish_initial.rds")
```

Visualize initial clusters
```{r}
# Zebrafish <- readRDS("../Species_Objects/Zebrafish_initial.rds")
DimPlot(Zebrafish, label = TRUE)
DimPlot(Zebrafish, group.by = "orig.file", cells = sample(colnames(Zebrafish)))
VlnPlot(Zebrafish, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Zebrafish, "nFeature_RNA", pt.size = 0) + RotatedAxis()
```

Plot canonical markers
```{r}
#Zebrafish <- readRDS("../Species_Initial/Zebrafish_initial.rds")
#Zebrafish <- DendroOrder(Zebrafish)
#Zebrafish <- UpperCase_genes(Zebrafish)

RGC_markers= c("RBPMS2B", "POU6F2", "THY1")
BC_markers=c("VSX2", "PRKCA", "OTX2","VSX1")
AC_markers=c("TFAP2A","GAD2", "SLC6A9")
HC_markers=c("ONECUT1", "CALB1", "TPM3")
Cone_markers=c("ARR3A","PDE6H")
Rod_markers=c("RHO")
MG_markers=c("GLULA", "APOEB")
MicroG_markers=c("C1QA", "C1QB")

DotPlot(Zebrafish, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Cone_markers, Rod_markers, MG_markers, MicroG_markers, "CLDN5A","CLDN5B", "IGFBP7"), group.by = "dendro_order") + RotatedAxis()
```

Calculate DE genes for cluster annotation
```{r}
Zebrafish_genes <- FindAllMarkers(Zebrafish)
saveRDS(Zebrafish_genes, file = "../Marker_Genes/Zebrafish_genes_initial.rds")
```

Explore cells without clear markers
```{r}
Zebrafish <- readRDS("../Species_Initial/Zebrafish_initial.rds")
Idents(Zebrafish) <- "seurat_clusters"

# 8 likely glia: top DE genes include SEPP1A, GLULA, RLBP1A
markers8 <- FindMarkers(Zebrafish, ident.1 = 8, only.pos = TRUE, assay = "RNA", max.cells.per.ident = 3000, features = unique(rownames(Zebrafish)))
markers8 <- markers8[order(markers8$avg_log2FC, decreasing = TRUE),]
markers8 <- markers8[order(markers8$p_val),]

# Cluster 9 - likely glia, top DE genes include GLULA. RLBP1A, CRABP1A
markers9 <- FindMarkers(Zebrafish, ident.1 = 9, only.pos = TRUE, assay = "RNA", max.cells.per.ident = 3000, features = unique(rownames(Zebrafish)))
markers9 <- markers9[order(markers9$avg_log2FC, decreasing = TRUE),]



markers18 <- FindMarkers(Zebrafish, ident.1 = 18, only.pos = TRUE, assay = "RNA", max.cells.per.ident = 5000, features = unique(rownames(Zebrafish)))

# Top DE genes for 27 are mitochondrial genes: annotate as other
markers27 <- FindMarkers(Zebrafish, ident.1 = 27, only.pos = TRUE, assay = "RNA", max.cells.per.ident = 5000, features = unique(rownames(Zebrafish)))
markers27 <- markers27[order(markers27$avg_log2FC, decreasing = TRUE),]


```


Annotate major cell classes
```{r}
Idents(Zebrafish) <- "seurat_clusters"
Zebrafish@meta.data$cell_class = "Other"

Zebrafish@meta.data[WhichCells(Zebrafish, idents =c(30,31,28,22,14,21,4,15,29,17,19,10,13,20,38,35,26,6)),]$cell_class = "RGC"
Zebrafish@meta.data[WhichCells(Zebrafish, idents = c(11,7,5,3,33,36)),]$cell_class = "BP"
Zebrafish@meta.data[WhichCells(Zebrafish, idents = c(23)),]$cell_class = "HC"
Zebrafish@meta.data[WhichCells(Zebrafish, idents = c(18, 8, 9)),]$cell_class = "MG"
Zebrafish@meta.data[WhichCells(Zebrafish, idents = c(43)),]$cell_class = "MicroG"

# Separate AC into Gaba and Gly 
Zfish_AC <- subset(Zebrafish, cells = WhichCells(Zebrafish, idents =  c(24)))
Zfish_AC <- ClusterSeurat(Zfish_AC)

DotPlot(Zfish_AC, features = c("GAD1A","GAD1B", "GAD2", "TFAP2A", "TFAP2B", "SLC6A9")) + RotatedAxis()

GABA_cells <- WhichCells(Zfish_AC, idents = c(0,1,2,5,6,7,8,9))
Gly_cells <- WhichCells(Zfish_AC, idents = c(4,3))

Zebrafish@meta.data[GABA_cells,]$cell_class = "GabaAC"
Zebrafish@meta.data[Gly_cells,]$cell_class = "GlyAC"

# Separate PR into rods and cones
Zfish_PR <- subset(Zebrafish, cells = WhichCells(Zebrafish, idents =  c(12)))
Zfish_PR <- ClusterSeurat(Zfish_PR)

Rod_markers <- c("RHO" ,"PDC", "NRL", "SAG", "GNAT1", "GNGT1")
Cone_markers <- c("ARR3A", "ARR3B", "PDE6H", "GNAT2","GNGT2","OPN1SW","OPN1MW")
DotPlot(Zfish_PR, features = c(Rod_markers, Cone_markers)) + RotatedAxis()

rod_cells <- WhichCells(Zfish_PR, idents = c(4,6,7))
cone_cells <- WhichCells(Zfish_PR, idents = 8)

Zebrafish@meta.data[rod_cells,]$cell_class = "Rod"
Zebrafish@meta.data[cone_cells,]$cell_class = "Cone"


# Most remaining clusters are low quality.
Zebrafish@meta.data[WhichCells(Zebrafish, idents = c(0,1,2,16,25,27,32,34,37,39:42,44)),]$cell_class = "Other"

Idents(Zebrafish) <- "cell_class"
DimPlot(Zebrafish, group.by = "cell_class")


saveRDS(Zebrafish@meta.data, "../Metadata/Zebrafish_metadata.rds")
saveRDS(Zebrafish, "../Species_Objects/Zebrafish_initial.rds")
```


# Bipolar cells
```{r}
data_dir <- "/Users/joshhahn/Google Drive File Stream/My Drive/shekharlab_data/projects/Zebrafish/Loom/Bipolar/"
ZfishBC1 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC1/possorted_genome_bam_OQDY2.loom"), engine = "hdf5r")
ZfishBC2 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC2/possorted_genome_bam_TIPBO.loom"), engine = "hdf5r")
ZfishBC3 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC3/possorted_genome_bam_S06NP.loom"), engine = "hdf5r")
ZfishBC4 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC4/possorted_genome_bam_S557J.loom"), engine = "hdf5r")
ZfishBC5 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC5/possorted_genome_bam_QFA52.loom"), engine = "hdf5r")
ZfishBC6 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC6/possorted_genome_bam_AC7O9.loom"), engine = "hdf5r")
ZfishBC7 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC7/possorted_genome_bam_7JH2I.loom"), engine = "hdf5r")
ZfishBC8 <- read.loom.matrices(file = paste0(data_dir, "ZfishBC8/possorted_genome_bam_MROMJ.loom"), engine = "hdf5r")
```

Add introns and exons and combine matrices
```{r}
ZfishBC1_mat <- ZfishBC1$spliced + ZfishBC1$unspliced
ZfishBC2_mat <- ZfishBC2$spliced + ZfishBC2$unspliced
ZfishBC3_mat <- ZfishBC3$spliced + ZfishBC3$unspliced
ZfishBC4_mat <- ZfishBC4$spliced + ZfishBC4$unspliced
ZfishBC5_mat <- ZfishBC5$spliced + ZfishBC5$unspliced
ZfishBC6_mat <- ZfishBC6$spliced + ZfishBC6$unspliced
ZfishBC7_mat <- ZfishBC7$spliced + ZfishBC7$unspliced
ZfishBC8_mat <- ZfishBC8$spliced + ZfishBC8$unspliced

Zebrafish_mat <- cbind(ZfishBC1_mat, ZfishBC2_mat, ZfishBC3_mat, ZfishBC4_mat, ZfishBC5_mat, ZfishBC6_mat, ZfishBC7_mat, ZfishBC8_mat)
```

Create Seurat object
```{r}
Zebrafish <- CreateSeuratObject(Zebrafish_mat, names.delim = ":", min.features = 350)

Zebrafish@meta.data[colnames(ZfishBC1_mat), 'orig.file'] = "ZfishBC1"
Zebrafish@meta.data[colnames(ZfishBC2_mat), 'orig.file'] = "ZfishBC2"
Zebrafish@meta.data[colnames(ZfishBC3_mat), 'orig.file'] = "ZfishBC3"
Zebrafish@meta.data[colnames(ZfishBC4_mat), 'orig.file'] = "ZfishBC4"
Zebrafish@meta.data[colnames(ZfishBC5_mat), 'orig.file'] = "ZfishBC5"
Zebrafish@meta.data[colnames(ZfishBC6_mat), 'orig.file'] = "ZfishBC6"
Zebrafish@meta.data[colnames(ZfishBC7_mat), 'orig.file'] = "ZfishBC7"
Zebrafish@meta.data[colnames(ZfishBC8_mat), 'orig.file'] = "ZfishBC8"
```

Initial quality control metrics
```{r}
VlnPlot(Zebrafish, features = "nCount_RNA", pt.size = 0)
VlnPlot(Zebrafish, features = "nFeature_RNA", pt.size = 0)
```

Perform initial clustering
```{r}
Zfish_BC <- subset(Zfish_BC, subset = nCount_RNA < 2500 & nFeature_RNA > 50)
Zfish_BC <- ClusterSeurat(Zfish_BC)
saveRDS(Zfish_BC, file = "../Species_Objects/Zebrafish_BC_initial.rds")
```

Visualize initial clusters
```{r}
Zfish_BC <- readRDS("../Species_Objects/Zebrafish_BC_initial (1).rds")
DimPlot(Zfish_BC, label = TRUE)
DimPlot(Zfish_BC, group.by = "orig.file", cells = sample(colnames(Zfish_BC)))
VlnPlot(Zfish_BC, "nCount_RNA", pt.size = 0) + RotatedAxis() + NoLegend()
VlnPlot(Zfish_BC, "nFeature_RNA", pt.size = 0) + RotatedAxis() + NoLegend()
```

Verify clusters correspond to bipolar cells
```{r}
RGC_markers= c("RBPMS2B", "POU6F2", "THY1", "SLC17A6", "ROBO2")
BC_markers=c("CABP5A", "VSX1","VSX2","CABP5", "OTX2", "PRKCA", "GRIK1")
AC_markers=c("TFAP2A","TFAP2B", "ONECUT2","GAD1", "GAD2", "SLC6A9")
HC_markers=c("ONECUT1", "CALB1")
PR_markers=c("PDE6H","ARR3A", "ARR3B", "RHO")
MG_markers=c("APOEA", "APOEB","GLULA","GLULB","GLULC","RLBP1A", "RLBP1B", "CRABP1A", "CRABP1B", "DKK3A", "DKK3B", "CRYM", "C1QA", "C1QB")

DotPlot(Zfish_BC, features = c(BC_markers, RGC_markers, AC_markers, HC_markers, PR_markers, MG_markers), assay = "RNA", group.by = "seurat_clusters") + RotatedAxis()
```

Calculate DE genes to see if any clusters should be removed / merged
```{r}
ZfishBC_markers <- FindAllMarkers(Zfish_BC, only.pos = TRUE, assay = "RNA")
saveRDS(ZfishBC_markers, file = "../Species_Markers/ZebrafishBCmarkers_initial.rds")
```

Plot top DE genes
```{r}
Zfish_BC <- readRDS("../Species_Objects/Zebrafish_BC_initial (1).rds")
ZfishBC_markers <- readRDS("../Species_Markers/ZebrafishBCmarkers_initial.rds")

Zfish_BC <- DendroOrder(Zfish_BC)
Idents(Zfish_BC) <- "dendro_order"

DotPlot(Zfish_BC, features = TopMarkers(Zfish_BC, markers = ZfishBC_markers, num_markers = 2), assay = "RNA", group.by = "dendro_order") + RotatedAxis()
```

Remove contaminant cell types
```{r}
Zfish_BC <- readRDS("../Species_Objects/Zebrafish_BC_initial (1).rds")
ZfishBC_markers <- readRDS("../Species_Markers/ZebrafishBCmarkers_v1.rds")

# Cluster 23 is also amacrine cells, SLC6A9, TFAP2A, TFAP2B
# Cluster 24 is amacrine cells
print("Cluster 24 DE genes:")
head(subset(ZfishBC_markers, cluster == 24), 10)$gene
# Cluster 25 is rods
print("Cluster 25 DE genes:")
head(subset(ZfishBC_markers, cluster == 25), 10)$gene
# Questionable: 3 and 5 express GLULA specifically, but retain bipolar markers
print("Cluster 3 DE genes:")
head(subset(ZfishBC_markers, cluster == 3), 10)$gene

Zfish_BC <- DropClusters(Zfish_BC, idents = c(23, 24,25), refactor = TRUE)
saveRDS(Zfish_BC, "../Species_Objects/ZebrafishBC_v2.rds")
```

Calculate DE  markers
```{r}
Zfish_BC <- readRDS("../Species_Objects/ZebrafishBC_v3.rds")

Idents(Zfish_BC) <- "seurat_clusters"
cluster_markers <- FindAllMarkers(Zfish_BC, assay = "RNA", only.pos = TRUE, test.use = "MAST")

saveRDS(cluster_markers, "../Species_Markers/ZebrafishBCmarkers_v3.rds")
```

```{r}
cluster_markers <- readRDS("../Species_Markers/ZebrafishBCmarkers_v3.rds")

DotPlot(Zfish_BC, features = TopMarkers(cluster_markers, num_markers = 2)) + RotatedAxis()


clus19_markers <- subset(cluster_markers, cluster == 19)
clus14_markers <- subset(cluster_markers, cluster == 14)


DotPlot(Zfish_BC, features = head(clus19_markers$gene, 20)) + RotatedAxis()
```


Convert UMAP and diagonal plot to figure files
Verify all cells are bipolar
```{r}
Zfish_BC <- readRDS("../Species_Objects/ZebrafishBC_v2.rds")

pdf("../Figures/Species_UMAPs/ZebrafishBC.pdf", w=4, h=4, useDingbats = FALSE)
DimPlot(Zfish_BC, label = TRUE) + NoLegend() + ggtitle("Zebrafish Bipolars")
dev.off()

RGC_markers= c("RBPMS2B", "ROBO2")
BC_markers=c("CABP5A", "VSX1")
AC_markers=c("TFAP2A","TFAP2B","GAD2", "SLC6A9")
HC_markers=c("ONECUT1", "ONECUT2")
PR_markers=c("PDE6H","ARR3A", "ARR3B", "RHO")

DotPlot(Zfish_BC, features = c(BC_markers, RGC_markers, AC_markers, HC_markers, PR_markers), assay = "RNA", group.by = "seurat_clusters") + RotatedAxis()

```

Annotate bipolar cells as rod, Cone On, Cone Off
```{r}
Zfish_BC <- readRDS("../Species_Objects/ZebrafishBC_v2.rds")

pan_BC <- c("VSX1", "CABP5A")
Rod <- c("PRKCA.1", "GRAMD1B")
Cone <- c("SLC24A3", "APBA2B")
Off <- c("GRIK1A", "GRIK1B")
On <- c("GRM6A","GRM6B", "ISL1", "GNB3A")

DotPlot(Zfish_BC, features = c(pan_BC, Rod, Cone, Off, On), assay = "RNA") + RotatedAxis()

```


```{r}
Idents(Zfish_BC) <- "seurat_clusters"

Zfish_BC@meta.data$bipolar_order = factor(Idents(Zfish_BC), levels = rev(c(14,19,1,7,8,9,16,17,18,22,23,2,3,4,5,6,10,11,12,13,15,20,21)))

Zfish_BC@meta.data$type <- "OFF Cone"
Zfish_BC@meta.data[WhichCells(Zfish_BC, idents = c(14,19)), "type"] <- "Rod Bipolar"
Zfish_BC@meta.data[WhichCells(Zfish_BC, idents = c(1,7,8,9,16,17,18,22,23)), "type"] <- "ON Cone"

saveRDS(Zfish_BC, "../Species_Objects/ZebrafishBC_v3.rds")
Zfish_BC <- readRDS("../Species_Objects/ZebrafishBC_v3.rds")
```

Access stability of dendrogram
```{r}
dendro_list <- list()
nfeatures <- c(100, 500, 1000, 2000)
for(i in 1:4){
  test <- FindVariableFeatures(Zfish_BC, selection.method = "vst", nfeatures = nfeatures[i])
  test <- BuildClusterTree(test)
  plot(test@tools$BuildClusterTree, main = paste0("Features: ", nfeatures[i]))
}
```


## Compare to Mouse bipolars
Extract Rod/Cone bipolar markers, ON/OFF markers from Mouse to compare to Zebrafish since the normal markers aren't sufficient
```{r}
Mouse_BC <- readRDS("../Species_Objects/MouseBC_integrated_v3.rds")
Idents(Mouse_BC) <- "annotated"
Mouse_BC <- subset(Mouse_BC, idents = setdiff(Idents(Mouse_BC), c("Rods", "GlyAC", "GabaAC", "MG")))
Idents(Mouse_BC) <- "annotated"

Mouse_BC@meta.data$Rod_Cone <- "Cone Bipolar"
Mouse_BC@meta.data[WhichCells(Mouse_BC, idents = "RBC"), "Rod_Cone"] <- "Rod Bipolar"

Mouse_BC@meta.data$ON_OFF <- "OFF"
Mouse_BC@meta.data[WhichCells(Mouse_BC, idents = c("BC5A", "BC5B", "BC5C", "BC5D", "BC6", "BC7", "BC8", "BC9", "RBC")), "ON_OFF"] <- "ON"

Idents(Mouse_BC) <- "Rod_Cone"
Rod_markers <- FindAllMarkers(Mouse_BC, only.pos = TRUE, assay = "RNA")

Idents(Mouse_BC) <- "ON_OFF"
ON_markers <- FindAllMarkers(Mouse_BC, only.pos = TRUE, assay = "RNA")
```

```{r}
RBC_mark <- head(rownames(subset(Rod_markers, cluster == "Rod Bipolar")), 20)
Cone_mark <- head(rownames(subset(Rod_markers, cluster == "Cone Bipolar")), 20)
DotPlot(Mouse_BC, features = c(RBC_mark, Cone_mark), group.by = "annotated", assay = "RNA") + RotatedAxis() + ggtitle("Mouse Bipolars")
DotPlot(Zfish_BC, features = c(RBC_mark, Cone_mark), assay = "RNA") + RotatedAxis()


ON_mark <- head(rownames(subset(markers, cluster == "ON Cone")), 20)
OFF_mark <- head(rownames(subset(markers, cluster == "OFF Cone")), 20)
DotPlot(Mouse_BC, features = c(ON_mark, OFF_mark), group.by = "annotated", assay = "RNA") + RotatedAxis()
DotPlot(Zfish_BC, features = c(ON_mark, OFF_mark), assay = "RNA") + RotatedAxis()
```

Tru supervised classification
```{r}
Mouse_ortho <- readRDS("../Species_OrthologMatrix/Mouse_orthomat_v3.rds")
Zfish_ortho <- readRDS("../Species_OrthologMatrix/Zebrafish_orthomat_v3.rds")

colnames(Mouse_ortho) <- substr(colnames(Mouse_ortho), 7, 50)
colnames(Zfish_ortho) <- substr(colnames(Zfish_ortho), 11, 54)

Mouse_ortho <- Mouse_ortho[,colnames(Mouse_ortho) %in% colnames(Mouse_BC)]
Zfish_ortho <- Zfish_ortho[,colnames(Zfish_ortho) %in% colnames(Zfish_BC)]

BC_mouse <- CreateSeuratObject(Mouse_ortho)
BC_Zfish <- CreateSeuratObject(Zfish_ortho)

BC_mouse <- ClusterSeurat(BC_mouse)
BC_Zfish <- ClusterSeurat(BC_Zfish)

BC_mouse@meta.data$annotated <- Mouse_BC@meta.data[colnames(BC_mouse), "annotated"]
BC_Zfish@meta.data$annotated <- Zfish_BC@meta.data[colnames(BC_Zfish), "seurat_clusters"]

saveRDS(BC_mouse, "../Species_OrthologMatrix/Mouse_BC_ortho_obj.rds")
saveRDS(BC_Zfish, "../Species_OrthologMatrix/Zebrafish_BC_ortho_obj.rds")
```

```{r}
BC_mouse <- readRDS("../Species_OrthologMatrix/Mouse_BC_ortho_obj.rds")
BC_Zfish <- readRDS("../Species_OrthologMatrix/Zebrafish_BC_ortho_obj.rds")

Zfish_HVG <- VariableFeatures(BC_Zfish)
Mouse_HVG <- VariableFeatures(BC_mouse)
training_features <- unique(c(Zfish_HVG, Mouse_HVG))

Idents(BC_mouse) <- "annotated"
Mouse_model <- TrainModel(BC_mouse, training_genes = training_features, train_ident = "annotated")

Idents(BC_Zfish) <- "annotated"
Z_M_mat <- BuildConfusionMatrix(test = BC_Zfish, train = BC_mouse, model = Mouse_model)
MakePrettyConfusionMatrix(Z_M_mat)
```

Ortholog plots of best Rod/ON Cone/OFF Cone markers
```{r}
BC_Zfish@meta.data$bipolar_order = factor(Idents(BC_Zfish), levels = rev(c(14,19,1,7,8,9,16,17,18,22,23,2,3,4,5,6,10,11,12,13,15,20,21)))


DotPlot(BC_mouse, features = c("Prkca","Gramd1b","Apba2", "Slc24a3","Grik1", "Isl1", "Gnb3"), assay = "RNA", group.by = "annotated") + RotatedAxis() + ggtitle("Mouse Bipolar")
DotPlot(BC_Zfish, features = c("Prkca","Gramd1b","Apba2", "Slc24a3","Grik1", "Isl1", "Gnb3"), assay = "RNA", group.by = "bipolar_order") + RotatedAxis() + ggtitle("Zebrafish Bipolar")
```

Calculate new ortholog markers for Rod/ON Cone/OFF Cone 
```{r}
Idents(BC_mouse) <- "annotated"
BC_mouse@meta.data$Rod_Cone <- "Cone Bipolar"
BC_mouse@meta.data[WhichCells(BC_mouse, idents = "RBC"), "Rod_Cone"] <- "Rod Bipolar"

BC_mouse@meta.data$ON_OFF <- "OFF"
BC_mouse@meta.data[WhichCells(BC_mouse, idents = c("BC5A", "BC5B", "BC5C", "BC5D", "BC6", "BC7", "BC8", "BC9", "RBC")), "ON_OFF"] <- "ON"

Idents(BC_mouse) <- "Rod_Cone"
Rod_markers <- FindAllMarkers(BC_mouse, only.pos = TRUE, assay = "RNA")

Idents(BC_mouse) <- "ON_OFF"
ON_markers <- FindAllMarkers(BC_mouse, only.pos = TRUE, assay = "RNA")
```

```{r}
RBC_mark <- head(rownames(subset(Rod_markers, cluster == "Rod Bipolar")), 50)
Cone_mark <- head(rownames(subset(Rod_markers, cluster == "Cone Bipolar")), 40)
ON_mark <- head(rownames(subset(ON_markers, cluster == "ON")), 40)
OFF_mark <- head(rownames(subset(ON_markers, cluster == "OFF")), 40)

DotPlot(BC_mouse, features = c(RBC_mark), group.by = "annotated", assay = "RNA") + RotatedAxis() + ggtitle("Mouse Bipolars")
DotPlot(BC_mouse, features = c( Cone_mark), group.by = "annotated", assay = "RNA") + RotatedAxis() + ggtitle("Mouse Bipolars")

DotPlot(BC_Zfish, features = c(RBC_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Zebrafish Bipolars")
DotPlot(BC_Zfish, features = c(Cone_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Zebrafish Bipolars")



DotPlot(BC_mouse, features = c(ON_mark, OFF_mark), group.by = "annotated", assay = "RNA") + RotatedAxis() + ggtitle("Mouse Bipolars") 
DotPlot(BC_Zfish, features = c(ON_mark, OFF_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Zebrafish Bipolars")

```

## Compare to Chicken bipolars
```{r}
Chicken_ortho <- readRDS("../Species_OrthologMatrix/Chicken_orthomat_v3.rds")
Chicken_BC <- readRDS("../Species_Objects/ChickenBC_v1.rds")

colnames(Chicken_ortho) <- substr(colnames(Chicken_ortho), 9, 55)

Chicken_ortho <- Chicken_ortho[,colnames(Chicken_ortho) %in% colnames(Chicken_BC)]

BC_Chicken <- CreateSeuratObject(Chicken_ortho)

BC_Chicken <- ClusterSeurat(BC_Chicken)

BC_Chicken@meta.data$annotated <- Chicken_BC@meta.data[colnames(BC_Chicken), "annotated"]

saveRDS(BC_Chicken, "../Species_OrthologMatrix/Chicken_BC_ortho_obj.rds")
```

```{r}
BC_Chicken <- readRDS("../Species_OrthologMatrix/Chicken_BC_ortho_obj.rds")

Idents(BC_Chicken) <- "annotated"
BC_Chicken@meta.data$Rod_Cone <- "Cone Bipolar"
BC_Chicken@meta.data[WhichCells(BC_Chicken, idents = "BP-19"), "Rod_Cone"] <- "Rod Bipolar"

BC_Chicken@meta.data$ON_OFF <- "OFF"
BC_Chicken@meta.data[WhichCells(BC_Chicken, idents = c("BP-3", "BP-9", "BP-22", "BP-4", "BP-14", "BP-18", "BP-19", "BP-13", "BP-21", "BP-1")), "ON_OFF"] <- "ON"
BC_Chicken@meta.data[WhichCells(BC_Chicken, idents = c("BP-10")), "ON_OFF"] <- "ON/OFF"

Idents(BC_Chicken) <- "Rod_Cone"
Rod_markers <- FindAllMarkers(BC_Chicken, only.pos = TRUE, assay = "RNA")

Idents(BC_Chicken) <- "ON_OFF"
ON_markers <- FindAllMarkers(BC_Chicken, only.pos = TRUE, assay = "RNA")
```

```{r}
Idents(BC_Chicken) <- "annotated"
BC_Chicken@meta.data$bipolar_order = factor(Idents(BC_Chicken), levels = paste0("BP-", c(17,16,20,15,11,12,7,6,8,5,2,10,1,21,13,18,14,4,22,9,3,19)))


RBC_mark <- head(rownames(subset(Rod_markers, cluster == "Rod Bipolar")), 50)
Cone_mark <- head(rownames(subset(Rod_markers, cluster == "Cone Bipolar")), 50)
ON_mark <- head(rownames(subset(ON_markers, cluster == "ON")), 50)
OFF_mark <- head(rownames(subset(ON_markers, cluster == "OFF")), 50)

DotPlot(BC_Chicken, features = c(RBC_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Chicken Rod Bipolar Markers")
DotPlot(BC_Chicken, features = c(Cone_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Chicken Cone Bipolar Markers")

DotPlot(BC_Zfish, features = c(RBC_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Zebrafish Rod Bipolar Markers")
DotPlot(BC_Zfish, features = c(Cone_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Zebrafish Cone Bipolar Markers")



DotPlot(BC_Chicken, features = c(ON_mark, OFF_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Chicken Bipolars") 
DotPlot(BC_Zfish, features = c(ON_mark, OFF_mark), group.by = "bipolar_order", assay = "RNA") + RotatedAxis() + ggtitle("Zebrafish Bipolars")

```

Supervised classification any good? :(
```{r}
Zfish_HVG <- VariableFeatures(BC_Zfish)
Chicken_HVG <- VariableFeatures(BC_Chicken)
training_features <- unique(c(Zfish_HVG, Chicken_HVG))

Idents(BC_Chicken) <- "annotated"
Chicken_model <- TrainModel(BC_Chicken, training_genes = training_features, train_ident = "annotated")

Idents(BC_Zfish) <- "annotated"
Z_C_mat <- BuildConfusionMatrix(test = BC_Zfish, train = BC_Chicken, model = Chicken_model)
MakePrettyConfusionMatrix(Z_C_mat)
```

## Cluster 19: Rod bipolar or not?
```{r}
cluster_markers <- readRDS("../Species_Markers/ZebrafishBCmarkers_v3.rds")

DotPlot(Zfish_BC, features = TopMarkers(cluster_markers, num_markers = 2)) + RotatedAxis()

clus19_markers <- subset(cluster_markers, cluster == 19)
clus14_markers <- subset(cluster_markers, cluster == 14)

mark19 <- head(clus19_markers$gene, 20)
mark14 <- head(clus14_markers$gene, 20)
common_mark <- intersect(mark19, mark14)

mark19 <- mark19[!(mark19 %in% common_mark)]
mark14 <- mark14[!(mark14 %in% common_mark)]

DotPlot(Zfish_BC, features = c(mark19, mark14, common_mark), assay = "RNA", group.by = "bipolar_order") + RotatedAxis()
```

```{r}
# Set testing parameters
greater_than = .4
less_than = .2
# Do a DE test with combined group 19 and 14, compare to combined group 14 and 7, a cluster proximal to 14 on the dendrogram
markers14_19 <- FindMarkers(Zfish_BC, ident.1 = c(14,19), assay = "RNA", test.use = "MAST")

# Sort genes that appear in each cluster, but do not appear in other clusters
markers14_19 <- markers14_19[order(-markers14_19$avg_log2FC), ]

markers_pass <- names(which(tfPercentExpression(Zfish_BC, clusID = 14, tf = rownames(markers14_19)) > greater_than))
markers_pass <- names(which(tfPercentExpression(Zfish_BC, clusID = 19, tf = markers_pass) > greater_than))
for(i in levels(Idents(Zfish_BC))){
  if(i == 14 || i == 19){}
  else{
    markers_pass <- names(which(tfPercentExpression(Zfish_BC, clusID = i, tf = markers_pass) < less_than))
  }
}

DotPlot(Zfish_BC, features = markers_pass, assay = "RNA", group.by = "bipolar_order") + RotatedAxis()
```

Repeat for clusters 14 and 7
```{r}
# Set testing parameters
greater_than = .4
less_than = .2
# Do a DE test with combined group 19 and 14, compare to combined group 14 and 7, a cluster proximal to 14 on the dendrogram
markers14_7 <- FindMarkers(Zfish_BC, ident.1 = c(14,7), assay = "RNA", test.use = "MAST")

# Sort genes that appear in each cluster, but do not appear in other clusters
markers14_7 <- markers14_7[order(-markers14_7$avg_log2FC), ]

markers_pass <- names(which(tfPercentExpression(Zfish_BC, clusID = 14, tf = rownames(markers14_7)) > greater_than))
markers_pass <- names(which(tfPercentExpression(Zfish_BC, clusID = 7, tf = markers_pass) > greater_than))

for(i in levels(Idents(Zfish_BC))){
  if(i == 14 || i == 7){}
  else{
    markers_pass <- names(which(tfPercentExpression(Zfish_BC, clusID = i, tf = markers_pass) < less_than))
  }
}

```

New dendrogram, based on correlation distance
```{r}
n = 2000


Zfish_BC <- FindVariableFeatures(Zfish_BC, nfeatures = n, selection.method = "vst")
feature_set <- VariableFeatures(Zfish_BC)

# Compute average gene expression matrix
data.avg <- AverageExpression(object = Zfish_BC, assays = "RNA", 
      features = feature_set, verbose = FALSE, return.seurat = TRUE)
data.avg <- as.data.frame(data.avg@assays$RNA@data)
  
# Compute hierarchical tree 
dendro <- pvclust(data.avg, method.hclust = "complete", method.dist = "correlation", nboot = 10000)

# Plot the tree
pdf("../Figures/Dendrograms/ZebrafishBC.pdf", w=6, h=6, useDingbats = FALSE)
plot(dendro, main = "Zebrafish Bipolars")
dev.off()

```

```{r}
Zfish_BC <- readRDS("../Species_Objects/ZebrafishBC_v3.rds")
```



# RGCs

Read in total velocyto object and subset to RGCs
```{r}
Zebrafish <- readRDS("../Species_Objects/Zebrafish_initial.rds")
Idents(Zebrafish) <- "orig.file"
Zfish_RGC <- subset(Zebrafish, idents = paste0("ZfishRGC", 1:15))
Zfish_RGC@meta.data$cell_class <- NULL
```

Import labels from study
```{r}
# Import cluster labels
RGC_ref <- readRDS("../Species_Reference/Zebrafish_RGC_final.rds")
meta <- data.frame(cluster = RGC_ref@meta.data$clusterID, row.names = rownames(RGC_ref@meta.data))

barcode <- function(string){
    code <- strsplit(string, c(":"))
    barcode <- strsplit(code[[1]][2], "x")
    return (paste0(barcode[[1]][1], "-1"))
}

new_names <- unlist(lapply(colnames(Zfish_RGC), FUN = barcode))

Zfish_RGC@meta.data$barcode <- paste0(Zfish_RGC@meta.data$orig.file, "_", new_names)
Zfish_RGC@meta.data$annotated <- meta[Zfish_RGC@meta.data$barcode,"cluster"]

# Remove cells that lack annotations
Zfish_RGC@meta.data$keep <- !is.na(Zfish_RGC@meta.data$annotated)
Idents(Zfish_RGC) <- "keep"
Zfish_RGC <- subset(Zfish_RGC, idents = TRUE)
Zfish_RGC@meta.data$keep <- NULL
Zfish_RGC@meta.data$cell_class <- NULL

saveRDS(Zfish_RGC, file = "../Species_Objects/ZebrafishRGC_v1.rds")
```

Run clustering pipeline to properly transform data and embed
```{r}
# Zfish_RGC <- readRDS("../Species_Objects/ZebrafishRGC_v1.rds")
Zfish_RGC <- ClusterSeurat(Zfish_RGC)
Idents(Zfish_RGC) <- "annotated"
saveRDS(Zfish_RGC, file = "../Species_Objects/ZfishRGC_v1.rds")
```

Visualize initial clusters
```{r}
 #Zfish_RGC <- readRDS("../Species_Objects/ZfishRGC_v1.rds")
DimPlot(Zfish_RGC, label = TRUE, group.by = "annotated")
DimPlot(Zfish_RGC, label = TRUE, group.by = "seurat_clusters")
DimPlot(Zfish_RGC, group.by = "orig.file", cells = sample(colnames(Zfish_RGC)))
VlnPlot(Zfish_RGC, "nCount_RNA", pt.size = 0) + RotatedAxis()
VlnPlot(Zfish_RGC, "nFeature_RNA", pt.size = 0) + RotatedAxis()
saveRDS(Zfish_RGC, "../Species_Objects/ZebrafishRGC_v1.rds")
```

Calculate DE genes using velocyto data
```{r}
Zfish_RGC <- readRDS("../Species_Objects/ZebrafishRGC_v1.rds")
Idents(Zfish_RGC) <- "annotated"
ZfishRGC_markers <- FindAllMarkers(Zfish_RGC, asssay = "RNA", test.use = "MAST")
saveRDS(ZfishRGC_markers, "../Species_Markers/ZebrafishRGCmarkers.rds")
```

Convert UMAP and diagonal gene plots to files
```{r}
Zfish_RGC <- readRDS("../Species_Objects/ZebrafishRGC_v1.rds")
Idents(Zfish_RGC) <- "annotated"

pdf("../Figures/Species_UMAPs/ZfishRGC.pdf", w=4, h=4, useDingbats = FALSE)
DimPlot(Zfish_RGC, label = TRUE) + NoLegend()
dev.off()

markers <- PlotUniqueMarkers(Zfish_RGC, RGC_markers, edits = TRUE)
pdf("Figures/Species_MarkerPlots/ZfishRGC.pdf", , w=6, h=4, useDingbats = FALSE)

dev.off()

```

# Amacrine Cells
```{r}
Zebrafish <- readRDS("../Species_Objects/Zebrafish_initial.rds")
Idents(Zebrafish) <- "cell_class"
ZfishAC <- subset(Zebrafish, idents = "AC")

ZfishAC@meta.data$file <- substring(ZfishAC@meta.data$orig.file, 6,6)
Idents(ZfishAC) <- "file"
AC_genes <- FindAllMarkers(ZfishAC, assay = "RNA")
```

# Rods
```{r}
Zebrafish <- readRDS("../Species_Objects/Zebrafish_initial.rds")

# Only one photoreceptor cluster was detected: recluster to separate rods and cones
Idents(Zebrafish) <- "cell_class"
Zebrafish_PR <- subset(Zebrafish, idents = "PR")
Zebrafish_PR <- NormalizeData(Zebrafish_PR)
Zebrafish_PR <- FindVariableFeatures(Zebrafish_PR)
Zebrafish_PR <- ScaleData(Zebrafish_PR)
Zebrafish_PR <- RunPCA(Zebrafish_PR)
Zebrafish_PR <- FindNeighbors(Zebrafish_PR, dims = 1:30)
Zebrafish_PR <- FindClusters(Zebrafish_PR, resolution = .5)


Rod_markers <- c("RHO" , "NRL", "GNAT1", "GNGT1")
Cone_markers <- c("PDE6H", "GNAT2")
DotPlot(Zebrafish_PR, features = c(Rod_markers, Cone_markers)) + RotatedAxis()

# Subset to rods, do preliminary clustering, and save
Zebrafish_Rods <- subset(Zebrafish_PR, idents = c(4,5,6))
Zebrafish_Rods <- ClusterSeurat(Zebrafish_Rods)
saveRDS(Zebrafish_Rods, file = "../Species_Objects/ZebrafishRods_v1.rds")
```


# Scratch
```{r}
Zfish <- subset(Zebrafish, idents =setdiff(0:44, c(0,1,216,25,27,32,34,37,39:42,44)))
Zfish <- ClusterSeurat(Zfish)
```

```{r}
RGC_markers= c("RBPMS2B", "POU6F2", "THY1")
BC_markers=c("VSX2", "PRKCA", "OTX2","VSX1")
AC_markers=c("TFAP2A","GAD2", "SLC6A9")
HC_markers=c("ONECUT1", "CALB1", "TPM3")
Cone_markers=c("ARR3A","PDE6H")
Rod_markers=c("RHO")
MG_markers=c("GLULA", "APOEB")
MicroG_markers=c("C1QA", "C1QB")

DotPlot(Zfish, features = c(RGC_markers, BC_markers, AC_markers, HC_markers, Cone_markers, Rod_markers, MG_markers, MicroG_markers, "CLDN5A","CLDN5B", "IGFBP7"), group.by = "seurat_clusters") + RotatedAxis()
```

Plot unique markers
```{r}
cluster_markers <- FindAllMarkers(Zfish_BC, assay = "RNA", only.pos = TRUE)

Idents(Zfish_BC) <- "bipolar_order"
top_marks <- TopMarkers(cluster_markers, num_markers = 2)
top_marks <- DiagonalizeGenes(top_marks, Zfish_BC)
DotPlot(Zfish_BC, features = top_marks) + RotatedAxis()
```

Particular markers
```{r}
DotPlot(Mouse_BC, features  = c("SCG2", "PCDH7"), assay = "RNA", group.by = "annotated") + ggtitle("Mouse Bipolars") + RotatedAxis()
DotPlot(Chicken_BC, features  = c("SCG2", "PCDH7"), assay = "RNA", group.by = "bipolar_order") + ggtitle("Chicken Bipolars") + RotatedAxis()
DotPlot(Zfish_BC, features  = c("SCG2B", "PCDH7B"), assay = "RNA", group.by = "bipolar_order") + ggtitle("Zebrafish Bipolars") + RotatedAxis()

```

Glutamate receptors
```{r}
GRM <- grep("^GRM", rownames(Zfish_BC), value = TRUE)
GRIK <- grep("^GRIK", rownames(Zfish_BC), value = TRUE)
GRIA <- grep("^GRIA", rownames(Zfish_BC), value = TRUE)
GRIN <- grep("^GRIN", rownames(Zfish_BC), value = TRUE)
all <- c(GRM, GRIK, GRIA, GRIN)
highly <- c("GRM1B","GRM6A", "GRM6B", "GRIK1A","GRIN1B", "GRIN2CA")
lowly <- all[!(all %in% highly)]

DotPlot(Zfish_BC, features = c(GRM, GRIK, GRIA, GRIN), assay = "RNA", group.by = "bipolar_order") + RotatedAxis()

DotPlot(Zfish_BC, features = highly, assay = "RNA", group.by = "bipolar_order") + RotatedAxis()
DotPlot(Zfish_BC, features = lowly, assay = "RNA", group.by = "bipolar_order") + RotatedAxis()




```




